// dataverse-ify/lib/dataverse-ify/__tests__/integration-tests/addtoqueue.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const SetupGlobalContext_1 = require("../../../webapi/SetupGlobalContext");
const MetadataCache_1 = require("../../../metadata/MetadataCache");
const Entity_1 = require("../../../types/Entity");
const __1 = require("../..");
const config = require("config");
const Letter_1 = require("../../../dataverse-gen/entities/Letter");
const Queue_1 = require("../../../dataverse-gen/entities/Queue");
const QueueItem_1 = require("../../../dataverse-gen/entities/QueueItem");
const AddToQueue_1 = require("../../../dataverse-gen/actions/AddToQueue");
describe("addtoqueue", () => {
    const configFile = config.get("nodewebapi");
    beforeAll(() => __awaiter(void 0, void 0, void 0, function* () {
        if (!configFile.runIntegrationTests)
            return;
        // Is this running inside NodeJS?
        if (typeof Xrm === "undefined") {
            try {
                // Set up the Node Xrm global context
                yield SetupGlobalContext_1.SetupGlobalContext();
            }
            catch (ex) {
                fail(ex);
            }
        }
    }), 300000);
    test("AddToQueue", () => __awaiter(void 0, void 0, void 0, function* () {
        if (!configFile.runIntegrationTests)
            return;
        MetadataCache_1.setMetadataCache({
            entities: {
                letter: Letter_1.letterMetadata,
                queue: Queue_1.queueMetadata,
                queueitem: QueueItem_1.queueitemMetadata,
            },
            actions: { AddToQueue: AddToQueue_1.AddToQueueMetadata },
        });
        const queue = {
            logicalName: Queue_1.queueMetadata.logicalName,
            name: "Sample Queue",
        };
        const letter = {
            logicalName: Letter_1.letterMetadata.logicalName,
            subject: "Sample Letter",
        };
        const cdsServiceClient = new __1.XrmContextCdsServiceClient(Xrm.WebApi);
        try {
            // Create Queue
            queue.id = yield cdsServiceClient.create(queue);
            // Create letter
            letter.id = yield cdsServiceClient.create(letter);
            // Add letter to queue
            const request = {
                logicalName: AddToQueue_1.AddToQueueMetadata.operationName,
                entity: Entity_1.Entity.toEntityReference(queue),
                Target: Entity_1.Entity.toEntityReference(letter),
            };
            const response = (yield cdsServiceClient.execute(request));
            expect(response.QueueItemId).toBeDefined();
        }
        catch (ex) {
            fail(ex);
        }
        finally {
            if (letter.id) {
                // Tidy up
                yield cdsServiceClient.delete(letter);
            }
            if (queue.id) {
                // Tidy up
                yield cdsServiceClient.delete(queue);
            }
        }
    }), 300000);
});
