// dataverse-ify/lib/dataverse-ify/__tests__/integration-tests/pascalcase-navigation-property.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable @typescript-eslint/camelcase */
const SetupGlobalContext_1 = require("../../../webapi/SetupGlobalContext");
const MetadataCache_1 = require("../../../metadata/MetadataCache");
const Account_1 = require("../../../dataverse-gen/entities/Account");
const __1 = require("../..");
const config = require("config");
const types_1 = require("../../../types");
describe("pascalcase-navigation-property", () => {
    const configFile = config.get("nodewebapi");
    beforeAll(() => __awaiter(void 0, void 0, void 0, function* () {
        if (!configFile.runIntegrationTests)
            return;
        // Is this running inside NodeJS?
        if (typeof Xrm === "undefined") {
            try {
                // Set up the Node Xrm global context
                yield SetupGlobalContext_1.SetupGlobalContext();
            }
            catch (ex) {
                fail(ex);
            }
        }
    }), 30000);
    test("pascalcase-navigation-property", () => __awaiter(void 0, void 0, void 0, function* () {
        if (!configFile.runIntegrationTests)
            return;
        MetadataCache_1.setMetadataCache({
            entities: {
                account: Account_1.accountMetadata,
            },
        });
        const cdsServiceClient = new __1.XrmContextCdsServiceClient(Xrm.WebApi);
        const accountRelated = {
            logicalName: Account_1.accountMetadata.logicalName,
            name: "Account " + new Date().toISOString(),
        };
        // Create account
        accountRelated.id = yield cdsServiceClient.create(accountRelated);
        const account1 = {
            logicalName: Account_1.accountMetadata.logicalName,
            name: "Account " + new Date().toISOString(),
            cdsify_account1: types_1.Entity.toEntityReference(accountRelated),
        };
        // Create account
        account1.id = yield cdsServiceClient.create(account1);
        // Retrieve the account
        const accountRetrieved = (yield cdsServiceClient.retrieve(account1.logicalName, account1.id, [
            Account_1.AccountAttributes.cdsify_Account1,
        ]));
        expect(accountRetrieved.cdsify_account1).toBeDefined();
        // Tidy up
        yield cdsServiceClient.delete(accountRelated);
        yield cdsServiceClient.delete(account1);
    }), 30000);
});
