// dataverse-ify/lib/dataverse-gen/metadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.metadataCache = exports.Entities = void 0;
/* eslint-disable*/
const Account_1 = require("./entities/Account");
const Contact_1 = require("./entities/Contact");
const Incident_1 = require("./entities/Incident");
const Letter_1 = require("./entities/Letter");
const Opportunity_1 = require("./entities/Opportunity");
const OpportunityClose_1 = require("./entities/OpportunityClose");
const Queue_1 = require("./entities/Queue");
const QueueItem_1 = require("./entities/QueueItem");
const AddToQueue_1 = require("./actions/AddToQueue");
const WinOpportunity_1 = require("./actions/WinOpportunity");
const CalculateRollupField_1 = require("./functions/CalculateRollupField");
const CalculateTotalTimeIncident_1 = require("./functions/CalculateTotalTimeIncident");
const RetrieveMetadataChanges_1 = require("./functions/RetrieveMetadataChanges");
exports.Entities = {
    Account: "account",
    Contact: "contact",
    Incident: "incident",
    Letter: "letter",
    Opportunity: "opportunity",
    OpportunityClose: "opportunityclose",
    Queue: "queue",
    QueueItem: "queueitem",
};
// Setup Metadata
// Usage: setMetadataCache(metadataCache);
exports.metadataCache = {
    entities: {
        account: Account_1.accountMetadata,
        contact: Contact_1.contactMetadata,
        incident: Incident_1.incidentMetadata,
        letter: Letter_1.letterMetadata,
        opportunity: Opportunity_1.opportunityMetadata,
        opportunityclose: OpportunityClose_1.opportunitycloseMetadata,
        queue: Queue_1.queueMetadata,
        queueitem: QueueItem_1.queueitemMetadata,
    },
    actions: {
        AddToQueue: AddToQueue_1.AddToQueueMetadata,
        WinOpportunity: WinOpportunity_1.WinOpportunityMetadata,
        CalculateRollupField: CalculateRollupField_1.CalculateRollupFieldMetadata,
        CalculateTotalTimeIncident: CalculateTotalTimeIncident_1.CalculateTotalTimeIncidentMetadata,
        RetrieveMetadataChanges: RetrieveMetadataChanges_1.RetrieveMetadataChangesMetadata,
    }
};
