// dataverse-ify/lib/dataverse-ify/__tests__/unit-tests/field-lookup.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable @typescript-eslint/camelcase */
const Account_1 = require("../../../dataverse-gen/entities/Account");
const metadata_1 = require("../../../metadata");
const odataify_1 = require("../../odataify/odataify");
const __1 = require("../../..");
test("odataify lookups", () => __awaiter(void 0, void 0, void 0, function* () {
    metadata_1.setMetadataCache({ entities: { account: Account_1.accountMetadata } });
    const accountSdk = {
        logicalName: Account_1.accountMetadata.logicalName,
        parentaccountid: new __1.EntityReference("account", "123"),
    };
    const accountOdata = yield odataify_1.odataify("Create", accountSdk);
    expect(accountOdata).toBeDefined();
    const expectedOdata = {
        "@odata.type": "Microsoft.Dynamics.CRM.account",
        "parentaccountid@odata.bind": "accounts(123)",
    };
    expect(JSON.stringify(accountOdata)).toBe(JSON.stringify(expectedOdata));
}));
test("odataify lookups", () => __awaiter(void 0, void 0, void 0, function* () {
    metadata_1.setMetadataCache({ entities: { account: Account_1.accountMetadata } });
    const odataRecord = {
        "@odata.context": "https://develop1v9demo.crm11.dynamics.com/api/data/v9.1/$metadata#accounts/$entity",
        // eslint-disable-next-line quotes
        "@odata.etag": 'W/"24588382"',
        name: "Account 1",
        "_preferredsystemuserid_value@OData.Community.Display.V1.FormattedValue": "Scott Durow",
        "_preferredsystemuserid_value@Microsoft.Dynamics.CRM.lookuplogicalname": "systemuser",
        _preferredsystemuserid_value: "d3b787e3-8c39-48e6-97fc-26dfc14f104d",
    };
    const sdkRecord = yield __1.sdkify(odataRecord, "account");
    const expectedSdk = {
        name: "Account 1",
        logicalName: "account",
        formattedValues: {
            _preferredsystemuserid_value: "Scott Durow",
        },
        preferredsystemuserid: {
            entityType: "systemuser",
            id: "d3b787e3-8c39-48e6-97fc-26dfc14f104d",
            name: "Scott Durow",
        },
    };
    expect(JSON.stringify(sdkRecord)).toBe(JSON.stringify(expectedSdk));
}));
