// dataverse-ify/lib/dataverse-gen/functions/RetrieveMetadataChanges.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RetrieveMetadataChangesMetadata = void 0;
const StructuralProperty_1 = require("../../types/StructuralProperty");
const OperationType_1 = require("../../types/OperationType");
// Action RetrieveMetadataChanges
exports.RetrieveMetadataChangesMetadata = {
    parameterTypes: {
        "Query": {
            typeName: "mscrm.EntityQueryExpression",
            structuralProperty: StructuralProperty_1.StructuralProperty.EntityType
        },
        "DeletedMetadataFilters": {
            typeName: "mscrm.DeletedMetadataFilters",
            structuralProperty: StructuralProperty_1.StructuralProperty.EnumerationType
        },
        "ClientVersionStamp": {
            typeName: "Edm.String",
            structuralProperty: StructuralProperty_1.StructuralProperty.PrimitiveType
        },
        "AppModuleId": {
            typeName: "Edm.Guid",
            structuralProperty: StructuralProperty_1.StructuralProperty.PrimitiveType
        },
        "RetrieveAllSettings": {
            typeName: "Edm.Boolean",
            structuralProperty: StructuralProperty_1.StructuralProperty.PrimitiveType
        },
    },
    operationType: OperationType_1.OperationType.Function,
    operationName: "RetrieveMetadataChanges"
};
