// dataverse-ify/lib/webapi/GlobalContextStatic.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlobalContextStatic = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
const XrmStatic_1 = require("./XrmStatic");
class GlobalContextStatic {
    constructor(userId) {
        this._userId = userId;
    }
    getAdvancedConfigSetting(_setting) {
        throw new Error("Method not implemented.");
    }
    getClientUrl() {
        // Return the full url of the endpoint in the form "https://org.crm11.dynamics.com"
        return XrmStatic_1.XrmStatic.webapiInstance.getClientUrl().toString();
    }
    getCurrentAppName() {
        throw new Error("Method not implemented.");
    }
    getCurrentAppProperties() {
        throw new Error("Method not implemented.");
    }
    getCurrentAppUrl() {
        throw new Error("Method not implemented.");
    }
    getCurrentTheme() {
        throw new Error("Method not implemented.");
    }
    getIsAutoSaveEnabled() {
        throw new Error("Method not implemented.");
    }
    getOrgLcid() {
        throw new Error("Method not implemented.");
    }
    getOrgUniqueName() {
        throw new Error("Method not implemented.");
    }
    getQueryStringParameters() {
        throw new Error("Method not implemented.");
    }
    getTimeZoneOffsetMinutes() {
        throw new Error("Method not implemented.");
    }
    getUserId() {
        return this._userId;
    }
    getUserLcid() {
        throw new Error("Method not implemented.");
    }
    getUserName() {
        throw new Error("Method not implemented.");
    }
    getUserRoles() {
        throw new Error("Method not implemented.");
    }
    getVersion() {
        return XrmStatic_1.XrmStatic.webapiInstance.apiVersion;
    }
    isOnPremise() {
        throw new Error("Method not implemented.");
    }
    prependOrgName(_path) {
        throw new Error("Method not implemented.");
    }
}
exports.GlobalContextStatic = GlobalContextStatic;
