// dataverse-ify/lib/webapi/whoAmI.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.whoAmI = void 0;
/* eslint-disable @typescript-eslint/no-explicit-any */
const OperationType_1 = require("../types/OperationType");
const Guid_1 = require("../types/Guid");
function whoAmI() {
    return __awaiter(this, void 0, void 0, function* () {
        const request = new (class {
            constructor() {
                this.Target = {};
            }
            getMetadata() {
                return {
                    parameterTypes: {},
                    operationType: OperationType_1.OperationType.Function,
                    operationName: "WhoAmI",
                };
            }
        })();
        const whoAmIResponse = yield Xrm.WebApi.online.execute(request);
        const response = yield whoAmIResponse.json();
        return Guid_1.trimGuid(response.UserId);
    });
}
exports.whoAmI = whoAmI;
