// dataverse-ify/lib/webapi/XrmUtilityStatic.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeXrmUtilityStatic = void 0;
class NodeXrmUtilityStatic {
    getLearningPathAttributeName() {
        throw new Error("Method not implemented.");
    }
    getPageContext() {
        throw new Error("Method not implemented.");
    }
    closeProgressIndicator() {
        throw new Error("Not implemented");
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    getAllowedStatusTransitions(_entityName, _stateCode) {
        throw new Error("Not implemented");
    }
    getEntityMetadata(entityName, attributes) {
        const webApi = Xrm.WebApi;
        return webApi.getEntityMetadata(entityName, attributes);
    }
    getGlobalContext() {
        throw new Error("Not implemented");
    }
    getResourceString(_webResourceName, _key) {
        throw new Error("Not implemented");
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    invokeProcessAction(_name, _parameters) {
        throw new Error("Not implemented");
    }
    lookupObjects(_lookupOptions) {
        throw new Error("Not implemented");
    }
    refreshParentGrid(_lookupOptions) {
        throw new Error("Not implemented");
    }
    showProgressIndicator(_message) {
        throw new Error("Not implemented");
    }
    alertDialog(_message, _onCloseCallback) {
        throw new Error("Not implemented");
    }
    confirmDialog(_message, _yesCloseCallback, _noCloseCallback) {
        throw new Error("Not implemented");
    }
    isActivityType(_entityType) {
        throw new Error("Not implemented");
    }
    openQuickCreate(_entityLogicalName, _createFromEntity, _parameters) {
        throw new Error("Not implemented");
    }
    openEntityForm(_name, _id, _parameters, _windowOptions) {
        throw new Error("Not implemented");
    }
    openWebResource(_webResourceName, _webResourceData, _width, _height) {
        throw new Error("Not implemented");
    }
}
exports.NodeXrmUtilityStatic = NodeXrmUtilityStatic;
