// dataverse-ify/lib/dataverse-gen/entities/Account.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountAttributes = exports.accountMetadata = void 0;
// Entity Account
exports.accountMetadata = {
    typeName: "mscrm.account",
    logicalName: "account",
    collectionName: "accounts",
    primaryIdAttribute: "accountid",
    attributeTypes: {
        // Numeric Types
        address1_latitude: "Double",
        address1_longitude: "Double",
        address1_utcoffset: "Integer",
        address2_latitude: "Double",
        address2_longitude: "Double",
        address2_utcoffset: "Integer",
        aging30: "Money",
        aging30_base: "Money",
        aging60: "Money",
        aging60_base: "Money",
        aging90: "Money",
        aging90_base: "Money",
        creditlimit: "Money",
        creditlimit_base: "Money",
        entityimage_timestamp: "BigInt",
        exchangerate: "Decimal",
        importsequencenumber: "Integer",
        marketcap: "Money",
        marketcap_base: "Money",
        numberofemployees: "Integer",
        onholdtime: "Integer",
        opendeals: "Integer",
        opendeals_state: "Integer",
        openrevenue: "Money",
        openrevenue_base: "Money",
        openrevenue_state: "Integer",
        revenue: "Money",
        revenue_base: "Money",
        sharesoutstanding: "Integer",
        teamsfollowed: "Integer",
        timezoneruleversionnumber: "Integer",
        utcconversiontimezonecode: "Integer",
        versionnumber: "BigInt",
        // Optionsets
        accountcategorycode: "Optionset",
        accountclassificationcode: "Optionset",
        accountratingcode: "Optionset",
        address1_addresstypecode: "Optionset",
        address1_freighttermscode: "Optionset",
        address1_shippingmethodcode: "Optionset",
        address2_addresstypecode: "Optionset",
        address2_freighttermscode: "Optionset",
        address2_shippingmethodcode: "Optionset",
        businesstypecode: "Optionset",
        cdsify_multiselect: "MultiSelect",
        customersizecode: "Optionset",
        customertypecode: "Optionset",
        industrycode: "Optionset",
        ownershipcode: "Optionset",
        paymenttermscode: "Optionset",
        preferredappointmentdaycode: "Optionset",
        preferredappointmenttimecode: "Optionset",
        preferredcontactmethodcode: "Optionset",
        shippingmethodcode: "Optionset",
        statecode: "Optionset",
        statuscode: "Optionset",
        territorycode: "Optionset",
        // Date Formats
        createdon: "DateAndTime:UserLocal",
        lastonholdtime: "DateAndTime:UserLocal",
        lastusedincampaign: "DateOnly:UserLocal",
        modifiedon: "DateAndTime:UserLocal",
        opendeals_date: "DateAndTime:UserLocal",
        openrevenue_date: "DateAndTime:UserLocal",
        overriddencreatedon: "DateOnly:UserLocal",
    },
    navigation: {
        transactioncurrencyid: ["mscrm.transactioncurrency"],
        territoryid: ["mscrm.territory"],
        stageid_processstage: ["mscrm.processstage"],
        sla_account_sla: ["mscrm.sla"],
        primarycontactid: ["mscrm.contact"],
        preferredsystemuserid: ["mscrm.systemuser"],
        preferredserviceid: ["mscrm.service"],
        preferredequipmentid: ["mscrm.equipment"],
        parentaccountid: ["mscrm.account"],
        owninguser: ["mscrm.systemuser"],
        owningteam: ["mscrm.team"],
        owningbusinessunit: ["mscrm.businessunit"],
        ownerid: ["mscrm.principal"],
        originatingleadid: ["mscrm.lead"],
        msdyn_accountkpiid: ["mscrm.msdyn_accountkpiitem"],
        modifiedonbehalfby: ["mscrm.systemuser"],
        modifiedby: ["mscrm.systemuser"],
        masterid: ["mscrm.account"],
        defaultpricelevelid: ["mscrm.pricelevel"],
        createdonbehalfby: ["mscrm.systemuser"],
        createdby: ["mscrm.systemuser"],
        cdsify_Account1: ["mscrm.account"],
        slainvokedid: ["sla"],
    },
};
// Attribute constants
var AccountAttributes;
(function (AccountAttributes) {
    AccountAttributes["AccountCategoryCode"] = "accountcategorycode";
    AccountAttributes["AccountClassificationCode"] = "accountclassificationcode";
    AccountAttributes["AccountId"] = "accountid";
    AccountAttributes["AccountNumber"] = "accountnumber";
    AccountAttributes["AccountRatingCode"] = "accountratingcode";
    AccountAttributes["Address1_AddressId"] = "address1_addressid";
    AccountAttributes["Address1_AddressTypeCode"] = "address1_addresstypecode";
    AccountAttributes["Address1_City"] = "address1_city";
    AccountAttributes["Address1_Composite"] = "address1_composite";
    AccountAttributes["Address1_Country"] = "address1_country";
    AccountAttributes["Address1_County"] = "address1_county";
    AccountAttributes["Address1_Fax"] = "address1_fax";
    AccountAttributes["Address1_FreightTermsCode"] = "address1_freighttermscode";
    AccountAttributes["Address1_Latitude"] = "address1_latitude";
    AccountAttributes["Address1_Line1"] = "address1_line1";
    AccountAttributes["Address1_Line2"] = "address1_line2";
    AccountAttributes["Address1_Line3"] = "address1_line3";
    AccountAttributes["Address1_Longitude"] = "address1_longitude";
    AccountAttributes["Address1_Name"] = "address1_name";
    AccountAttributes["Address1_PostalCode"] = "address1_postalcode";
    AccountAttributes["Address1_PostOfficeBox"] = "address1_postofficebox";
    AccountAttributes["Address1_PrimaryContactName"] = "address1_primarycontactname";
    AccountAttributes["Address1_ShippingMethodCode"] = "address1_shippingmethodcode";
    AccountAttributes["Address1_StateOrProvince"] = "address1_stateorprovince";
    AccountAttributes["Address1_Telephone1"] = "address1_telephone1";
    AccountAttributes["Address1_Telephone2"] = "address1_telephone2";
    AccountAttributes["Address1_Telephone3"] = "address1_telephone3";
    AccountAttributes["Address1_UPSZone"] = "address1_upszone";
    AccountAttributes["Address1_UTCOffset"] = "address1_utcoffset";
    AccountAttributes["Address2_AddressId"] = "address2_addressid";
    AccountAttributes["Address2_AddressTypeCode"] = "address2_addresstypecode";
    AccountAttributes["Address2_City"] = "address2_city";
    AccountAttributes["Address2_Composite"] = "address2_composite";
    AccountAttributes["Address2_Country"] = "address2_country";
    AccountAttributes["Address2_County"] = "address2_county";
    AccountAttributes["Address2_Fax"] = "address2_fax";
    AccountAttributes["Address2_FreightTermsCode"] = "address2_freighttermscode";
    AccountAttributes["Address2_Latitude"] = "address2_latitude";
    AccountAttributes["Address2_Line1"] = "address2_line1";
    AccountAttributes["Address2_Line2"] = "address2_line2";
    AccountAttributes["Address2_Line3"] = "address2_line3";
    AccountAttributes["Address2_Longitude"] = "address2_longitude";
    AccountAttributes["Address2_Name"] = "address2_name";
    AccountAttributes["Address2_PostalCode"] = "address2_postalcode";
    AccountAttributes["Address2_PostOfficeBox"] = "address2_postofficebox";
    AccountAttributes["Address2_PrimaryContactName"] = "address2_primarycontactname";
    AccountAttributes["Address2_ShippingMethodCode"] = "address2_shippingmethodcode";
    AccountAttributes["Address2_StateOrProvince"] = "address2_stateorprovince";
    AccountAttributes["Address2_Telephone1"] = "address2_telephone1";
    AccountAttributes["Address2_Telephone2"] = "address2_telephone2";
    AccountAttributes["Address2_Telephone3"] = "address2_telephone3";
    AccountAttributes["Address2_UPSZone"] = "address2_upszone";
    AccountAttributes["Address2_UTCOffset"] = "address2_utcoffset";
    AccountAttributes["Aging30"] = "aging30";
    AccountAttributes["Aging30_Base"] = "aging30_base";
    AccountAttributes["Aging60"] = "aging60";
    AccountAttributes["Aging60_Base"] = "aging60_base";
    AccountAttributes["Aging90"] = "aging90";
    AccountAttributes["Aging90_Base"] = "aging90_base";
    AccountAttributes["BusinessTypeCode"] = "businesstypecode";
    AccountAttributes["cdsify_Account1"] = "cdsify_account1";
    AccountAttributes["cdsify_Account1Name"] = "cdsify_account1name";
    AccountAttributes["cdsify_Account1YomiName"] = "cdsify_account1yominame";
    AccountAttributes["cdsify_MultiSelect"] = "cdsify_multiselect";
    AccountAttributes["CreatedBy"] = "createdby";
    AccountAttributes["CreatedByExternalParty"] = "createdbyexternalparty";
    AccountAttributes["CreatedByExternalPartyName"] = "createdbyexternalpartyname";
    AccountAttributes["CreatedByExternalPartyYomiName"] = "createdbyexternalpartyyominame";
    AccountAttributes["CreatedByName"] = "createdbyname";
    AccountAttributes["CreatedByYomiName"] = "createdbyyominame";
    AccountAttributes["CreatedOn"] = "createdon";
    AccountAttributes["CreatedOnBehalfBy"] = "createdonbehalfby";
    AccountAttributes["CreatedOnBehalfByName"] = "createdonbehalfbyname";
    AccountAttributes["CreatedOnBehalfByYomiName"] = "createdonbehalfbyyominame";
    AccountAttributes["CreditLimit"] = "creditlimit";
    AccountAttributes["CreditLimit_Base"] = "creditlimit_base";
    AccountAttributes["CreditOnHold"] = "creditonhold";
    AccountAttributes["CustomerSizeCode"] = "customersizecode";
    AccountAttributes["CustomerTypeCode"] = "customertypecode";
    AccountAttributes["DefaultPriceLevelId"] = "defaultpricelevelid";
    AccountAttributes["DefaultPriceLevelIdName"] = "defaultpricelevelidname";
    AccountAttributes["Description"] = "description";
    AccountAttributes["DoNotBulkEMail"] = "donotbulkemail";
    AccountAttributes["DoNotBulkPostalMail"] = "donotbulkpostalmail";
    AccountAttributes["DoNotEMail"] = "donotemail";
    AccountAttributes["DoNotFax"] = "donotfax";
    AccountAttributes["DoNotPhone"] = "donotphone";
    AccountAttributes["DoNotPostalMail"] = "donotpostalmail";
    AccountAttributes["DoNotSendMM"] = "donotsendmm";
    AccountAttributes["EMailAddress1"] = "emailaddress1";
    AccountAttributes["EMailAddress2"] = "emailaddress2";
    AccountAttributes["EMailAddress3"] = "emailaddress3";
    AccountAttributes["EntityImage"] = "entityimage";
    AccountAttributes["EntityImage_Timestamp"] = "entityimage_timestamp";
    AccountAttributes["EntityImage_URL"] = "entityimage_url";
    AccountAttributes["EntityImageId"] = "entityimageid";
    AccountAttributes["ExchangeRate"] = "exchangerate";
    AccountAttributes["Fax"] = "fax";
    AccountAttributes["FollowEmail"] = "followemail";
    AccountAttributes["FtpSiteURL"] = "ftpsiteurl";
    AccountAttributes["ImportSequenceNumber"] = "importsequencenumber";
    AccountAttributes["IndustryCode"] = "industrycode";
    AccountAttributes["IsPrivate"] = "isprivate";
    AccountAttributes["LastOnHoldTime"] = "lastonholdtime";
    AccountAttributes["LastUsedInCampaign"] = "lastusedincampaign";
    AccountAttributes["MarketCap"] = "marketcap";
    AccountAttributes["MarketCap_Base"] = "marketcap_base";
    AccountAttributes["MarketingOnly"] = "marketingonly";
    AccountAttributes["MasterAccountIdName"] = "masteraccountidname";
    AccountAttributes["MasterAccountIdYomiName"] = "masteraccountidyominame";
    AccountAttributes["MasterId"] = "masterid";
    AccountAttributes["Merged"] = "merged";
    AccountAttributes["ModifiedBy"] = "modifiedby";
    AccountAttributes["ModifiedByExternalParty"] = "modifiedbyexternalparty";
    AccountAttributes["ModifiedByExternalPartyName"] = "modifiedbyexternalpartyname";
    AccountAttributes["ModifiedByExternalPartyYomiName"] = "modifiedbyexternalpartyyominame";
    AccountAttributes["ModifiedByName"] = "modifiedbyname";
    AccountAttributes["ModifiedByYomiName"] = "modifiedbyyominame";
    AccountAttributes["ModifiedOn"] = "modifiedon";
    AccountAttributes["ModifiedOnBehalfBy"] = "modifiedonbehalfby";
    AccountAttributes["ModifiedOnBehalfByName"] = "modifiedonbehalfbyname";
    AccountAttributes["ModifiedOnBehalfByYomiName"] = "modifiedonbehalfbyyominame";
    AccountAttributes["msdyn_accountkpiid"] = "msdyn_accountkpiid";
    AccountAttributes["msdyn_accountkpiidName"] = "msdyn_accountkpiidname";
    AccountAttributes["msdyn_gdproptout"] = "msdyn_gdproptout";
    AccountAttributes["Name"] = "name";
    AccountAttributes["NumberOfEmployees"] = "numberofemployees";
    AccountAttributes["OnHoldTime"] = "onholdtime";
    AccountAttributes["OpenDeals"] = "opendeals";
    AccountAttributes["OpenDeals_Date"] = "opendeals_date";
    AccountAttributes["OpenDeals_State"] = "opendeals_state";
    AccountAttributes["OpenRevenue"] = "openrevenue";
    AccountAttributes["OpenRevenue_Base"] = "openrevenue_base";
    AccountAttributes["OpenRevenue_Date"] = "openrevenue_date";
    AccountAttributes["OpenRevenue_State"] = "openrevenue_state";
    AccountAttributes["OriginatingLeadId"] = "originatingleadid";
    AccountAttributes["OriginatingLeadIdName"] = "originatingleadidname";
    AccountAttributes["OriginatingLeadIdYomiName"] = "originatingleadidyominame";
    AccountAttributes["OverriddenCreatedOn"] = "overriddencreatedon";
    AccountAttributes["OwnerId"] = "ownerid";
    AccountAttributes["OwnerIdName"] = "owneridname";
    AccountAttributes["OwnerIdType"] = "owneridtype";
    AccountAttributes["OwnerIdYomiName"] = "owneridyominame";
    AccountAttributes["OwnershipCode"] = "ownershipcode";
    AccountAttributes["OwningBusinessUnit"] = "owningbusinessunit";
    AccountAttributes["OwningBusinessUnitName"] = "owningbusinessunitname";
    AccountAttributes["OwningTeam"] = "owningteam";
    AccountAttributes["OwningUser"] = "owninguser";
    AccountAttributes["ParentAccountId"] = "parentaccountid";
    AccountAttributes["ParentAccountIdName"] = "parentaccountidname";
    AccountAttributes["ParentAccountIdYomiName"] = "parentaccountidyominame";
    AccountAttributes["ParticipatesInWorkflow"] = "participatesinworkflow";
    AccountAttributes["PaymentTermsCode"] = "paymenttermscode";
    AccountAttributes["PreferredAppointmentDayCode"] = "preferredappointmentdaycode";
    AccountAttributes["PreferredAppointmentTimeCode"] = "preferredappointmenttimecode";
    AccountAttributes["PreferredContactMethodCode"] = "preferredcontactmethodcode";
    AccountAttributes["PreferredEquipmentId"] = "preferredequipmentid";
    AccountAttributes["PreferredEquipmentIdName"] = "preferredequipmentidname";
    AccountAttributes["PreferredServiceId"] = "preferredserviceid";
    AccountAttributes["PreferredServiceIdName"] = "preferredserviceidname";
    AccountAttributes["PreferredSystemUserId"] = "preferredsystemuserid";
    AccountAttributes["PreferredSystemUserIdName"] = "preferredsystemuseridname";
    AccountAttributes["PreferredSystemUserIdYomiName"] = "preferredsystemuseridyominame";
    AccountAttributes["PrimaryContactId"] = "primarycontactid";
    AccountAttributes["PrimaryContactIdName"] = "primarycontactidname";
    AccountAttributes["PrimaryContactIdYomiName"] = "primarycontactidyominame";
    AccountAttributes["PrimarySatoriId"] = "primarysatoriid";
    AccountAttributes["PrimaryTwitterId"] = "primarytwitterid";
    AccountAttributes["ProcessId"] = "processid";
    AccountAttributes["Revenue"] = "revenue";
    AccountAttributes["Revenue_Base"] = "revenue_base";
    AccountAttributes["SharesOutstanding"] = "sharesoutstanding";
    AccountAttributes["ShippingMethodCode"] = "shippingmethodcode";
    AccountAttributes["SIC"] = "sic";
    AccountAttributes["SLAId"] = "slaid";
    AccountAttributes["SLAInvokedId"] = "slainvokedid";
    AccountAttributes["SLAInvokedIdName"] = "slainvokedidname";
    AccountAttributes["SLAName"] = "slaname";
    AccountAttributes["StageId"] = "stageid";
    AccountAttributes["StateCode"] = "statecode";
    AccountAttributes["StatusCode"] = "statuscode";
    AccountAttributes["StockExchange"] = "stockexchange";
    AccountAttributes["TeamsFollowed"] = "teamsfollowed";
    AccountAttributes["Telephone1"] = "telephone1";
    AccountAttributes["Telephone2"] = "telephone2";
    AccountAttributes["Telephone3"] = "telephone3";
    AccountAttributes["TerritoryCode"] = "territorycode";
    AccountAttributes["TerritoryId"] = "territoryid";
    AccountAttributes["TerritoryIdName"] = "territoryidname";
    AccountAttributes["TickerSymbol"] = "tickersymbol";
    AccountAttributes["TimeSpentByMeOnEmailAndMeetings"] = "timespentbymeonemailandmeetings";
    AccountAttributes["TimeZoneRuleVersionNumber"] = "timezoneruleversionnumber";
    AccountAttributes["TransactionCurrencyId"] = "transactioncurrencyid";
    AccountAttributes["TransactionCurrencyIdName"] = "transactioncurrencyidname";
    AccountAttributes["TraversedPath"] = "traversedpath";
    AccountAttributes["UTCConversionTimeZoneCode"] = "utcconversiontimezonecode";
    AccountAttributes["VersionNumber"] = "versionnumber";
    AccountAttributes["WebSiteURL"] = "websiteurl";
    AccountAttributes["YomiName"] = "yominame";
})(AccountAttributes = exports.AccountAttributes || (exports.AccountAttributes = {}));
