// dataverse-ify/lib/dataverse-gen/entities/Contact.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContactAttributes = exports.contactMetadata = void 0;
// Entity Contact
exports.contactMetadata = {
    typeName: "mscrm.contact",
    logicalName: "contact",
    collectionName: "contacts",
    primaryIdAttribute: "contactid",
    attributeTypes: {
        // Numeric Types
        address1_latitude: "Double",
        address1_longitude: "Double",
        address1_utcoffset: "Integer",
        address2_latitude: "Double",
        address2_longitude: "Double",
        address2_utcoffset: "Integer",
        address3_latitude: "Double",
        address3_longitude: "Double",
        address3_utcoffset: "Integer",
        aging30: "Money",
        aging30_base: "Money",
        aging60: "Money",
        aging60_base: "Money",
        aging90: "Money",
        aging90_base: "Money",
        annualincome: "Money",
        annualincome_base: "Money",
        creditlimit: "Money",
        creditlimit_base: "Money",
        entityimage_timestamp: "BigInt",
        exchangerate: "Decimal",
        importsequencenumber: "Integer",
        numberofchildren: "Integer",
        onholdtime: "Integer",
        teamsfollowed: "Integer",
        timezoneruleversionnumber: "Integer",
        utcconversiontimezonecode: "Integer",
        versionnumber: "BigInt",
        // Optionsets
        accountrolecode: "Optionset",
        address1_addresstypecode: "Optionset",
        address1_freighttermscode: "Optionset",
        address1_shippingmethodcode: "Optionset",
        address2_addresstypecode: "Optionset",
        address2_freighttermscode: "Optionset",
        address2_shippingmethodcode: "Optionset",
        address3_addresstypecode: "Optionset",
        address3_freighttermscode: "Optionset",
        address3_shippingmethodcode: "Optionset",
        customersizecode: "Optionset",
        customertypecode: "Optionset",
        educationcode: "Optionset",
        familystatuscode: "Optionset",
        gendercode: "Optionset",
        haschildrencode: "Optionset",
        leadsourcecode: "Optionset",
        msdyn_orgchangestatus: "Optionset",
        paymenttermscode: "Optionset",
        preferredappointmentdaycode: "Optionset",
        preferredappointmenttimecode: "Optionset",
        preferredcontactmethodcode: "Optionset",
        shippingmethodcode: "Optionset",
        statecode: "Optionset",
        statuscode: "Optionset",
        territorycode: "Optionset",
        // Date Formats
        anniversary: "DateOnly:DateOnly",
        birthdate: "DateOnly:DateOnly",
        createdon: "DateAndTime:UserLocal",
        lastonholdtime: "DateAndTime:UserLocal",
        lastusedincampaign: "DateOnly:UserLocal",
        modifiedon: "DateAndTime:UserLocal",
        overriddencreatedon: "DateOnly:UserLocal",
    },
    navigation: {
        transactioncurrencyid: ["mscrm.transactioncurrency"],
        stageid_processstage: ["mscrm.processstage"],
        sla_contact_sla: ["mscrm.sla"],
        preferredsystemuserid: ["mscrm.systemuser"],
        preferredserviceid: ["mscrm.service"],
        preferredequipmentid: ["mscrm.equipment"],
        owninguser: ["mscrm.systemuser"],
        owningteam: ["mscrm.team"],
        owningbusinessunit: ["mscrm.businessunit"],
        ownerid: ["mscrm.principal"],
        originatingleadid: ["mscrm.lead"],
        msdyn_contactkpiid: ["mscrm.msdyn_contactkpiitem"],
        modifiedonbehalfby: ["mscrm.systemuser"],
        modifiedby: ["mscrm.systemuser"],
        masterid: ["mscrm.contact"],
        defaultpricelevelid: ["mscrm.pricelevel"],
        createdonbehalfby: ["mscrm.systemuser"],
        createdby: ["mscrm.systemuser"],
        parentcustomerid: ["account", "contact"],
        slainvokedid: ["sla"],
    },
};
// Attribute constants
var ContactAttributes;
(function (ContactAttributes) {
    ContactAttributes["AccountId"] = "accountid";
    ContactAttributes["AccountIdName"] = "accountidname";
    ContactAttributes["AccountIdYomiName"] = "accountidyominame";
    ContactAttributes["AccountRoleCode"] = "accountrolecode";
    ContactAttributes["Address1_AddressId"] = "address1_addressid";
    ContactAttributes["Address1_AddressTypeCode"] = "address1_addresstypecode";
    ContactAttributes["Address1_City"] = "address1_city";
    ContactAttributes["Address1_Composite"] = "address1_composite";
    ContactAttributes["Address1_Country"] = "address1_country";
    ContactAttributes["Address1_County"] = "address1_county";
    ContactAttributes["Address1_Fax"] = "address1_fax";
    ContactAttributes["Address1_FreightTermsCode"] = "address1_freighttermscode";
    ContactAttributes["Address1_Latitude"] = "address1_latitude";
    ContactAttributes["Address1_Line1"] = "address1_line1";
    ContactAttributes["Address1_Line2"] = "address1_line2";
    ContactAttributes["Address1_Line3"] = "address1_line3";
    ContactAttributes["Address1_Longitude"] = "address1_longitude";
    ContactAttributes["Address1_Name"] = "address1_name";
    ContactAttributes["Address1_PostalCode"] = "address1_postalcode";
    ContactAttributes["Address1_PostOfficeBox"] = "address1_postofficebox";
    ContactAttributes["Address1_PrimaryContactName"] = "address1_primarycontactname";
    ContactAttributes["Address1_ShippingMethodCode"] = "address1_shippingmethodcode";
    ContactAttributes["Address1_StateOrProvince"] = "address1_stateorprovince";
    ContactAttributes["Address1_Telephone1"] = "address1_telephone1";
    ContactAttributes["Address1_Telephone2"] = "address1_telephone2";
    ContactAttributes["Address1_Telephone3"] = "address1_telephone3";
    ContactAttributes["Address1_UPSZone"] = "address1_upszone";
    ContactAttributes["Address1_UTCOffset"] = "address1_utcoffset";
    ContactAttributes["Address2_AddressId"] = "address2_addressid";
    ContactAttributes["Address2_AddressTypeCode"] = "address2_addresstypecode";
    ContactAttributes["Address2_City"] = "address2_city";
    ContactAttributes["Address2_Composite"] = "address2_composite";
    ContactAttributes["Address2_Country"] = "address2_country";
    ContactAttributes["Address2_County"] = "address2_county";
    ContactAttributes["Address2_Fax"] = "address2_fax";
    ContactAttributes["Address2_FreightTermsCode"] = "address2_freighttermscode";
    ContactAttributes["Address2_Latitude"] = "address2_latitude";
    ContactAttributes["Address2_Line1"] = "address2_line1";
    ContactAttributes["Address2_Line2"] = "address2_line2";
    ContactAttributes["Address2_Line3"] = "address2_line3";
    ContactAttributes["Address2_Longitude"] = "address2_longitude";
    ContactAttributes["Address2_Name"] = "address2_name";
    ContactAttributes["Address2_PostalCode"] = "address2_postalcode";
    ContactAttributes["Address2_PostOfficeBox"] = "address2_postofficebox";
    ContactAttributes["Address2_PrimaryContactName"] = "address2_primarycontactname";
    ContactAttributes["Address2_ShippingMethodCode"] = "address2_shippingmethodcode";
    ContactAttributes["Address2_StateOrProvince"] = "address2_stateorprovince";
    ContactAttributes["Address2_Telephone1"] = "address2_telephone1";
    ContactAttributes["Address2_Telephone2"] = "address2_telephone2";
    ContactAttributes["Address2_Telephone3"] = "address2_telephone3";
    ContactAttributes["Address2_UPSZone"] = "address2_upszone";
    ContactAttributes["Address2_UTCOffset"] = "address2_utcoffset";
    ContactAttributes["Address3_AddressId"] = "address3_addressid";
    ContactAttributes["Address3_AddressTypeCode"] = "address3_addresstypecode";
    ContactAttributes["Address3_City"] = "address3_city";
    ContactAttributes["Address3_Composite"] = "address3_composite";
    ContactAttributes["Address3_Country"] = "address3_country";
    ContactAttributes["Address3_County"] = "address3_county";
    ContactAttributes["Address3_Fax"] = "address3_fax";
    ContactAttributes["Address3_FreightTermsCode"] = "address3_freighttermscode";
    ContactAttributes["Address3_Latitude"] = "address3_latitude";
    ContactAttributes["Address3_Line1"] = "address3_line1";
    ContactAttributes["Address3_Line2"] = "address3_line2";
    ContactAttributes["Address3_Line3"] = "address3_line3";
    ContactAttributes["Address3_Longitude"] = "address3_longitude";
    ContactAttributes["Address3_Name"] = "address3_name";
    ContactAttributes["Address3_PostalCode"] = "address3_postalcode";
    ContactAttributes["Address3_PostOfficeBox"] = "address3_postofficebox";
    ContactAttributes["Address3_PrimaryContactName"] = "address3_primarycontactname";
    ContactAttributes["Address3_ShippingMethodCode"] = "address3_shippingmethodcode";
    ContactAttributes["Address3_StateOrProvince"] = "address3_stateorprovince";
    ContactAttributes["Address3_Telephone1"] = "address3_telephone1";
    ContactAttributes["Address3_Telephone2"] = "address3_telephone2";
    ContactAttributes["Address3_Telephone3"] = "address3_telephone3";
    ContactAttributes["Address3_UPSZone"] = "address3_upszone";
    ContactAttributes["Address3_UTCOffset"] = "address3_utcoffset";
    ContactAttributes["Aging30"] = "aging30";
    ContactAttributes["Aging30_Base"] = "aging30_base";
    ContactAttributes["Aging60"] = "aging60";
    ContactAttributes["Aging60_Base"] = "aging60_base";
    ContactAttributes["Aging90"] = "aging90";
    ContactAttributes["Aging90_Base"] = "aging90_base";
    ContactAttributes["Anniversary"] = "anniversary";
    ContactAttributes["AnnualIncome"] = "annualincome";
    ContactAttributes["AnnualIncome_Base"] = "annualincome_base";
    ContactAttributes["AssistantName"] = "assistantname";
    ContactAttributes["AssistantPhone"] = "assistantphone";
    ContactAttributes["BirthDate"] = "birthdate";
    ContactAttributes["Business2"] = "business2";
    ContactAttributes["BusinessCard"] = "businesscard";
    ContactAttributes["BusinessCardAttributes"] = "businesscardattributes";
    ContactAttributes["Callback"] = "callback";
    ContactAttributes["ChildrensNames"] = "childrensnames";
    ContactAttributes["Company"] = "company";
    ContactAttributes["ContactId"] = "contactid";
    ContactAttributes["CreatedBy"] = "createdby";
    ContactAttributes["CreatedByExternalParty"] = "createdbyexternalparty";
    ContactAttributes["CreatedByExternalPartyName"] = "createdbyexternalpartyname";
    ContactAttributes["CreatedByExternalPartyYomiName"] = "createdbyexternalpartyyominame";
    ContactAttributes["CreatedByName"] = "createdbyname";
    ContactAttributes["CreatedByYomiName"] = "createdbyyominame";
    ContactAttributes["CreatedOn"] = "createdon";
    ContactAttributes["CreatedOnBehalfBy"] = "createdonbehalfby";
    ContactAttributes["CreatedOnBehalfByName"] = "createdonbehalfbyname";
    ContactAttributes["CreatedOnBehalfByYomiName"] = "createdonbehalfbyyominame";
    ContactAttributes["CreditLimit"] = "creditlimit";
    ContactAttributes["CreditLimit_Base"] = "creditlimit_base";
    ContactAttributes["CreditOnHold"] = "creditonhold";
    ContactAttributes["CustomerSizeCode"] = "customersizecode";
    ContactAttributes["CustomerTypeCode"] = "customertypecode";
    ContactAttributes["DefaultPriceLevelId"] = "defaultpricelevelid";
    ContactAttributes["DefaultPriceLevelIdName"] = "defaultpricelevelidname";
    ContactAttributes["Department"] = "department";
    ContactAttributes["Description"] = "description";
    ContactAttributes["DoNotBulkEMail"] = "donotbulkemail";
    ContactAttributes["DoNotBulkPostalMail"] = "donotbulkpostalmail";
    ContactAttributes["DoNotEMail"] = "donotemail";
    ContactAttributes["DoNotFax"] = "donotfax";
    ContactAttributes["DoNotPhone"] = "donotphone";
    ContactAttributes["DoNotPostalMail"] = "donotpostalmail";
    ContactAttributes["DoNotSendMM"] = "donotsendmm";
    ContactAttributes["EducationCode"] = "educationcode";
    ContactAttributes["EMailAddress1"] = "emailaddress1";
    ContactAttributes["EMailAddress2"] = "emailaddress2";
    ContactAttributes["EMailAddress3"] = "emailaddress3";
    ContactAttributes["EmployeeId"] = "employeeid";
    ContactAttributes["EntityImage"] = "entityimage";
    ContactAttributes["EntityImage_Timestamp"] = "entityimage_timestamp";
    ContactAttributes["EntityImage_URL"] = "entityimage_url";
    ContactAttributes["EntityImageId"] = "entityimageid";
    ContactAttributes["ExchangeRate"] = "exchangerate";
    ContactAttributes["ExternalUserIdentifier"] = "externaluseridentifier";
    ContactAttributes["FamilyStatusCode"] = "familystatuscode";
    ContactAttributes["Fax"] = "fax";
    ContactAttributes["FirstName"] = "firstname";
    ContactAttributes["FollowEmail"] = "followemail";
    ContactAttributes["FtpSiteUrl"] = "ftpsiteurl";
    ContactAttributes["FullName"] = "fullname";
    ContactAttributes["GenderCode"] = "gendercode";
    ContactAttributes["GovernmentId"] = "governmentid";
    ContactAttributes["HasChildrenCode"] = "haschildrencode";
    ContactAttributes["Home2"] = "home2";
    ContactAttributes["ImportSequenceNumber"] = "importsequencenumber";
    ContactAttributes["IsAutoCreate"] = "isautocreate";
    ContactAttributes["IsBackofficeCustomer"] = "isbackofficecustomer";
    ContactAttributes["IsPrivate"] = "isprivate";
    ContactAttributes["JobTitle"] = "jobtitle";
    ContactAttributes["LastName"] = "lastname";
    ContactAttributes["LastOnHoldTime"] = "lastonholdtime";
    ContactAttributes["LastUsedInCampaign"] = "lastusedincampaign";
    ContactAttributes["LeadSourceCode"] = "leadsourcecode";
    ContactAttributes["ManagerName"] = "managername";
    ContactAttributes["ManagerPhone"] = "managerphone";
    ContactAttributes["MarketingOnly"] = "marketingonly";
    ContactAttributes["MasterContactIdName"] = "mastercontactidname";
    ContactAttributes["MasterContactIdYomiName"] = "mastercontactidyominame";
    ContactAttributes["MasterId"] = "masterid";
    ContactAttributes["Merged"] = "merged";
    ContactAttributes["MiddleName"] = "middlename";
    ContactAttributes["MobilePhone"] = "mobilephone";
    ContactAttributes["ModifiedBy"] = "modifiedby";
    ContactAttributes["ModifiedByExternalParty"] = "modifiedbyexternalparty";
    ContactAttributes["ModifiedByExternalPartyName"] = "modifiedbyexternalpartyname";
    ContactAttributes["ModifiedByExternalPartyYomiName"] = "modifiedbyexternalpartyyominame";
    ContactAttributes["ModifiedByName"] = "modifiedbyname";
    ContactAttributes["ModifiedByYomiName"] = "modifiedbyyominame";
    ContactAttributes["ModifiedOn"] = "modifiedon";
    ContactAttributes["ModifiedOnBehalfBy"] = "modifiedonbehalfby";
    ContactAttributes["ModifiedOnBehalfByName"] = "modifiedonbehalfbyname";
    ContactAttributes["ModifiedOnBehalfByYomiName"] = "modifiedonbehalfbyyominame";
    ContactAttributes["msdyn_contactkpiid"] = "msdyn_contactkpiid";
    ContactAttributes["msdyn_contactkpiidName"] = "msdyn_contactkpiidname";
    ContactAttributes["msdyn_gdproptout"] = "msdyn_gdproptout";
    ContactAttributes["msdyn_orgchangestatus"] = "msdyn_orgchangestatus";
    ContactAttributes["NickName"] = "nickname";
    ContactAttributes["NumberOfChildren"] = "numberofchildren";
    ContactAttributes["OnHoldTime"] = "onholdtime";
    ContactAttributes["OriginatingLeadId"] = "originatingleadid";
    ContactAttributes["OriginatingLeadIdName"] = "originatingleadidname";
    ContactAttributes["OriginatingLeadIdYomiName"] = "originatingleadidyominame";
    ContactAttributes["OverriddenCreatedOn"] = "overriddencreatedon";
    ContactAttributes["OwnerId"] = "ownerid";
    ContactAttributes["OwnerIdName"] = "owneridname";
    ContactAttributes["OwnerIdType"] = "owneridtype";
    ContactAttributes["OwnerIdYomiName"] = "owneridyominame";
    ContactAttributes["OwningBusinessUnit"] = "owningbusinessunit";
    ContactAttributes["OwningBusinessUnitName"] = "owningbusinessunitname";
    ContactAttributes["OwningTeam"] = "owningteam";
    ContactAttributes["OwningUser"] = "owninguser";
    ContactAttributes["Pager"] = "pager";
    ContactAttributes["ParentContactId"] = "parentcontactid";
    ContactAttributes["ParentContactIdName"] = "parentcontactidname";
    ContactAttributes["ParentContactIdYomiName"] = "parentcontactidyominame";
    ContactAttributes["ParentCustomerId"] = "parentcustomerid";
    ContactAttributes["ParentCustomerIdName"] = "parentcustomeridname";
    ContactAttributes["ParentCustomerIdType"] = "parentcustomeridtype";
    ContactAttributes["ParentCustomerIdYomiName"] = "parentcustomeridyominame";
    ContactAttributes["ParticipatesInWorkflow"] = "participatesinworkflow";
    ContactAttributes["PaymentTermsCode"] = "paymenttermscode";
    ContactAttributes["PreferredAppointmentDayCode"] = "preferredappointmentdaycode";
    ContactAttributes["PreferredAppointmentTimeCode"] = "preferredappointmenttimecode";
    ContactAttributes["PreferredContactMethodCode"] = "preferredcontactmethodcode";
    ContactAttributes["PreferredEquipmentId"] = "preferredequipmentid";
    ContactAttributes["PreferredEquipmentIdName"] = "preferredequipmentidname";
    ContactAttributes["PreferredServiceId"] = "preferredserviceid";
    ContactAttributes["PreferredServiceIdName"] = "preferredserviceidname";
    ContactAttributes["PreferredSystemUserId"] = "preferredsystemuserid";
    ContactAttributes["PreferredSystemUserIdName"] = "preferredsystemuseridname";
    ContactAttributes["PreferredSystemUserIdYomiName"] = "preferredsystemuseridyominame";
    ContactAttributes["ProcessId"] = "processid";
    ContactAttributes["Salutation"] = "salutation";
    ContactAttributes["ShippingMethodCode"] = "shippingmethodcode";
    ContactAttributes["SLAId"] = "slaid";
    ContactAttributes["SLAInvokedId"] = "slainvokedid";
    ContactAttributes["SLAInvokedIdName"] = "slainvokedidname";
    ContactAttributes["SLAName"] = "slaname";
    ContactAttributes["SpousesName"] = "spousesname";
    ContactAttributes["StageId"] = "stageid";
    ContactAttributes["StateCode"] = "statecode";
    ContactAttributes["StatusCode"] = "statuscode";
    ContactAttributes["SubscriptionId"] = "subscriptionid";
    ContactAttributes["Suffix"] = "suffix";
    ContactAttributes["TeamsFollowed"] = "teamsfollowed";
    ContactAttributes["Telephone1"] = "telephone1";
    ContactAttributes["Telephone2"] = "telephone2";
    ContactAttributes["Telephone3"] = "telephone3";
    ContactAttributes["TerritoryCode"] = "territorycode";
    ContactAttributes["TimeSpentByMeOnEmailAndMeetings"] = "timespentbymeonemailandmeetings";
    ContactAttributes["TimeZoneRuleVersionNumber"] = "timezoneruleversionnumber";
    ContactAttributes["TransactionCurrencyId"] = "transactioncurrencyid";
    ContactAttributes["TransactionCurrencyIdName"] = "transactioncurrencyidname";
    ContactAttributes["TraversedPath"] = "traversedpath";
    ContactAttributes["UTCConversionTimeZoneCode"] = "utcconversiontimezonecode";
    ContactAttributes["VersionNumber"] = "versionnumber";
    ContactAttributes["WebSiteUrl"] = "websiteurl";
    ContactAttributes["YomiFirstName"] = "yomifirstname";
    ContactAttributes["YomiFullName"] = "yomifullname";
    ContactAttributes["YomiLastName"] = "yomilastname";
    ContactAttributes["YomiMiddleName"] = "yomimiddlename";
})(ContactAttributes = exports.ContactAttributes || (exports.ContactAttributes = {}));
