// dataverse-ify/lib/dataverse-gen/entities/Incident.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IncidentAttributes = exports.incidentMetadata = void 0;
// Entity Incident
exports.incidentMetadata = {
    typeName: "mscrm.incident",
    logicalName: "incident",
    collectionName: "incidents",
    primaryIdAttribute: "incidentid",
    attributeTypes: {
        // Numeric Types
        actualserviceunits: "Integer",
        billedserviceunits: "Integer",
        entityimage_timestamp: "BigInt",
        exchangerate: "Decimal",
        importsequencenumber: "Integer",
        influencescore: "Double",
        numberofchildincidents: "Integer",
        onholdtime: "Integer",
        sentimentvalue: "Double",
        timezoneruleversionnumber: "Integer",
        utcconversiontimezonecode: "Integer",
        versionnumber: "BigInt",
        // Optionsets
        caseorigincode: "Optionset",
        casetypecode: "Optionset",
        contractservicelevelcode: "Optionset",
        customersatisfactioncode: "Optionset",
        firstresponseslastatus: "Optionset",
        incidentstagecode: "Optionset",
        messagetypecode: "Optionset",
        prioritycode: "Optionset",
        resolvebyslastatus: "Optionset",
        servicestage: "Optionset",
        severitycode: "Optionset",
        statecode: "Optionset",
        statuscode: "Optionset",
        // Date Formats
        createdon: "DateAndTime:UserLocal",
        escalatedon: "DateAndTime:UserLocal",
        followupby: "DateOnly:UserLocal",
        lastonholdtime: "DateAndTime:UserLocal",
        modifiedon: "DateAndTime:UserLocal",
        overriddencreatedon: "DateOnly:UserLocal",
        resolveby: "DateAndTime:UserLocal",
        responseby: "DateAndTime:UserLocal",
    },
    navigation: {
        transactioncurrencyid: ["mscrm.transactioncurrency"],
        subjectid: ["mscrm.subject"],
        stageid_processstage: ["mscrm.processstage"],
        socialprofileid: ["mscrm.socialprofile"],
        responsiblecontactid: ["mscrm.contact"],
        resolvebykpiid: ["mscrm.slakpiinstance"],
        productid: ["mscrm.product"],
        primarycontactid: ["mscrm.contact"],
        parentcaseid: ["mscrm.incident"],
        owninguser: ["mscrm.systemuser"],
        owningteam: ["mscrm.team"],
        owningbusinessunit: ["mscrm.businessunit"],
        ownerid: ["mscrm.principal"],
        msdyn_IoTAlert: ["mscrm.msdyn_iotalert"],
        modifiedonbehalfby: ["mscrm.systemuser"],
        modifiedby: ["mscrm.systemuser"],
        masterid: ["mscrm.incident"],
        kbarticleid: ["mscrm.kbarticle"],
        firstresponsebykpiid: ["mscrm.slakpiinstance"],
        existingcase: ["mscrm.incident"],
        entitlementid: ["mscrm.entitlement"],
        createdonbehalfby: ["mscrm.systemuser"],
        createdby: ["mscrm.systemuser"],
        contractid: ["mscrm.contract"],
        contractdetailid: ["mscrm.contractdetail"],
        customerid: ["account", "contact"],
        slaid: ["sla"],
        slainvokedid: ["sla"],
    },
};
// Attribute constants
var IncidentAttributes;
(function (IncidentAttributes) {
    IncidentAttributes["AccountId"] = "accountid";
    IncidentAttributes["AccountIdName"] = "accountidname";
    IncidentAttributes["AccountIdYomiName"] = "accountidyominame";
    IncidentAttributes["ActivitiesComplete"] = "activitiescomplete";
    IncidentAttributes["ActualServiceUnits"] = "actualserviceunits";
    IncidentAttributes["BilledServiceUnits"] = "billedserviceunits";
    IncidentAttributes["BlockedProfile"] = "blockedprofile";
    IncidentAttributes["CaseOriginCode"] = "caseorigincode";
    IncidentAttributes["CaseTypeCode"] = "casetypecode";
    IncidentAttributes["CheckEmail"] = "checkemail";
    IncidentAttributes["ContactId"] = "contactid";
    IncidentAttributes["ContactIdName"] = "contactidname";
    IncidentAttributes["ContactIdYomiName"] = "contactidyominame";
    IncidentAttributes["ContractDetailId"] = "contractdetailid";
    IncidentAttributes["ContractDetailIdName"] = "contractdetailidname";
    IncidentAttributes["ContractId"] = "contractid";
    IncidentAttributes["ContractIdName"] = "contractidname";
    IncidentAttributes["ContractServiceLevelCode"] = "contractservicelevelcode";
    IncidentAttributes["CreatedBy"] = "createdby";
    IncidentAttributes["CreatedByExternalParty"] = "createdbyexternalparty";
    IncidentAttributes["CreatedByExternalPartyName"] = "createdbyexternalpartyname";
    IncidentAttributes["CreatedByExternalPartyYomiName"] = "createdbyexternalpartyyominame";
    IncidentAttributes["CreatedByName"] = "createdbyname";
    IncidentAttributes["CreatedByYomiName"] = "createdbyyominame";
    IncidentAttributes["CreatedOn"] = "createdon";
    IncidentAttributes["CreatedOnBehalfBy"] = "createdonbehalfby";
    IncidentAttributes["CreatedOnBehalfByName"] = "createdonbehalfbyname";
    IncidentAttributes["CreatedOnBehalfByYomiName"] = "createdonbehalfbyyominame";
    IncidentAttributes["CustomerContacted"] = "customercontacted";
    IncidentAttributes["CustomerId"] = "customerid";
    IncidentAttributes["CustomerIdName"] = "customeridname";
    IncidentAttributes["CustomerIdType"] = "customeridtype";
    IncidentAttributes["CustomerIdYomiName"] = "customeridyominame";
    IncidentAttributes["CustomerSatisfactionCode"] = "customersatisfactioncode";
    IncidentAttributes["DecrementEntitlementTerm"] = "decremententitlementterm";
    IncidentAttributes["Description"] = "description";
    IncidentAttributes["EmailAddress"] = "emailaddress";
    IncidentAttributes["EntitlementId"] = "entitlementid";
    IncidentAttributes["EntitlementIdName"] = "entitlementidname";
    IncidentAttributes["EntityImage"] = "entityimage";
    IncidentAttributes["EntityImage_Timestamp"] = "entityimage_timestamp";
    IncidentAttributes["EntityImage_URL"] = "entityimage_url";
    IncidentAttributes["EntityImageId"] = "entityimageid";
    IncidentAttributes["EscalatedOn"] = "escalatedon";
    IncidentAttributes["ExchangeRate"] = "exchangerate";
    IncidentAttributes["ExistingCase"] = "existingcase";
    IncidentAttributes["FirstResponseByKPIId"] = "firstresponsebykpiid";
    IncidentAttributes["FirstResponseByKPIIdName"] = "firstresponsebykpiidname";
    IncidentAttributes["FirstResponseSent"] = "firstresponsesent";
    IncidentAttributes["FirstResponseSLAStatus"] = "firstresponseslastatus";
    IncidentAttributes["FollowupBy"] = "followupby";
    IncidentAttributes["FollowUpTaskCreated"] = "followuptaskcreated";
    IncidentAttributes["ImportSequenceNumber"] = "importsequencenumber";
    IncidentAttributes["IncidentId"] = "incidentid";
    IncidentAttributes["IncidentStageCode"] = "incidentstagecode";
    IncidentAttributes["InfluenceScore"] = "influencescore";
    IncidentAttributes["IsDecrementing"] = "isdecrementing";
    IncidentAttributes["IsEscalated"] = "isescalated";
    IncidentAttributes["KbArticleId"] = "kbarticleid";
    IncidentAttributes["KbArticleIdName"] = "kbarticleidname";
    IncidentAttributes["LastOnHoldTime"] = "lastonholdtime";
    IncidentAttributes["MasterId"] = "masterid";
    IncidentAttributes["MasterIdName"] = "masteridname";
    IncidentAttributes["Merged"] = "merged";
    IncidentAttributes["MessageTypeCode"] = "messagetypecode";
    IncidentAttributes["ModifiedBy"] = "modifiedby";
    IncidentAttributes["ModifiedByExternalParty"] = "modifiedbyexternalparty";
    IncidentAttributes["ModifiedByExternalPartyName"] = "modifiedbyexternalpartyname";
    IncidentAttributes["ModifiedByExternalPartyYomiName"] = "modifiedbyexternalpartyyominame";
    IncidentAttributes["ModifiedByName"] = "modifiedbyname";
    IncidentAttributes["ModifiedByYomiName"] = "modifiedbyyominame";
    IncidentAttributes["ModifiedOn"] = "modifiedon";
    IncidentAttributes["ModifiedOnBehalfBy"] = "modifiedonbehalfby";
    IncidentAttributes["ModifiedOnBehalfByName"] = "modifiedonbehalfbyname";
    IncidentAttributes["ModifiedOnBehalfByYomiName"] = "modifiedonbehalfbyyominame";
    IncidentAttributes["msdyn_iotalert"] = "msdyn_iotalert";
    IncidentAttributes["msdyn_iotalertName"] = "msdyn_iotalertname";
    IncidentAttributes["NumberOfChildIncidents"] = "numberofchildincidents";
    IncidentAttributes["OnHoldTime"] = "onholdtime";
    IncidentAttributes["OverriddenCreatedOn"] = "overriddencreatedon";
    IncidentAttributes["OwnerId"] = "ownerid";
    IncidentAttributes["OwnerIdName"] = "owneridname";
    IncidentAttributes["OwnerIdType"] = "owneridtype";
    IncidentAttributes["OwnerIdYomiName"] = "owneridyominame";
    IncidentAttributes["OwningBusinessUnit"] = "owningbusinessunit";
    IncidentAttributes["OwningBusinessUnitName"] = "owningbusinessunitname";
    IncidentAttributes["OwningTeam"] = "owningteam";
    IncidentAttributes["OwningUser"] = "owninguser";
    IncidentAttributes["ParentCaseId"] = "parentcaseid";
    IncidentAttributes["ParentCaseIdName"] = "parentcaseidname";
    IncidentAttributes["PrimaryContactId"] = "primarycontactid";
    IncidentAttributes["PrimaryContactIdName"] = "primarycontactidname";
    IncidentAttributes["PrimaryContactIdYomiName"] = "primarycontactidyominame";
    IncidentAttributes["PriorityCode"] = "prioritycode";
    IncidentAttributes["ProcessId"] = "processid";
    IncidentAttributes["ProductId"] = "productid";
    IncidentAttributes["ProductIdName"] = "productidname";
    IncidentAttributes["ProductSerialNumber"] = "productserialnumber";
    IncidentAttributes["ResolveBy"] = "resolveby";
    IncidentAttributes["ResolveByKPIId"] = "resolvebykpiid";
    IncidentAttributes["ResolveByKPIIdName"] = "resolvebykpiidname";
    IncidentAttributes["ResolveBySLAStatus"] = "resolvebyslastatus";
    IncidentAttributes["ResponseBy"] = "responseby";
    IncidentAttributes["ResponsibleContactId"] = "responsiblecontactid";
    IncidentAttributes["ResponsibleContactIdName"] = "responsiblecontactidname";
    IncidentAttributes["ResponsibleContactIdYomiName"] = "responsiblecontactidyominame";
    IncidentAttributes["RouteCase"] = "routecase";
    IncidentAttributes["SentimentValue"] = "sentimentvalue";
    IncidentAttributes["ServiceStage"] = "servicestage";
    IncidentAttributes["SeverityCode"] = "severitycode";
    IncidentAttributes["SLAId"] = "slaid";
    IncidentAttributes["SLAInvokedId"] = "slainvokedid";
    IncidentAttributes["SLAInvokedIdName"] = "slainvokedidname";
    IncidentAttributes["SLAName"] = "slaname";
    IncidentAttributes["SocialProfileId"] = "socialprofileid";
    IncidentAttributes["SocialProfileIdName"] = "socialprofileidname";
    IncidentAttributes["StageId"] = "stageid";
    IncidentAttributes["StateCode"] = "statecode";
    IncidentAttributes["StatusCode"] = "statuscode";
    IncidentAttributes["SubjectId"] = "subjectid";
    IncidentAttributes["SubjectIdName"] = "subjectidname";
    IncidentAttributes["TicketNumber"] = "ticketnumber";
    IncidentAttributes["TimeZoneRuleVersionNumber"] = "timezoneruleversionnumber";
    IncidentAttributes["Title"] = "title";
    IncidentAttributes["TransactionCurrencyId"] = "transactioncurrencyid";
    IncidentAttributes["TransactionCurrencyIdName"] = "transactioncurrencyidname";
    IncidentAttributes["TraversedPath"] = "traversedpath";
    IncidentAttributes["UTCConversionTimeZoneCode"] = "utcconversiontimezonecode";
    IncidentAttributes["VersionNumber"] = "versionnumber";
})(IncidentAttributes = exports.IncidentAttributes || (exports.IncidentAttributes = {}));
