// dataverse-ify/lib/dataverse-gen/entities/Letter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LetterAttributes = exports.letterMetadata = void 0;
// Entity Letter
exports.letterMetadata = {
    typeName: "mscrm.letter",
    logicalName: "letter",
    collectionName: "letters",
    primaryIdAttribute: "activityid",
    attributeTypes: {
        // Numeric Types
        actualdurationminutes: "Integer",
        exchangerate: "Decimal",
        importsequencenumber: "Integer",
        onholdtime: "Integer",
        scheduleddurationminutes: "Integer",
        timezoneruleversionnumber: "Integer",
        utcconversiontimezonecode: "Integer",
        versionnumber: "BigInt",
        // Optionsets
        prioritycode: "Optionset",
        statecode: "Optionset",
        statuscode: "Optionset",
        // Date Formats
        actualend: "DateOnly:UserLocal",
        actualstart: "DateOnly:UserLocal",
        createdon: "DateAndTime:UserLocal",
        lastonholdtime: "DateAndTime:UserLocal",
        modifiedon: "DateAndTime:UserLocal",
        overriddencreatedon: "DateOnly:UserLocal",
        scheduledend: "DateAndTime:UserLocal",
        scheduledstart: "DateAndTime:UserLocal",
        sortdate: "DateAndTime:UserLocal",
    },
    navigation: {
        stageid_processstage: ["mscrm.processstage"],
        sla_letter_sla: ["mscrm.sla"],
        ownerid_letter: ["mscrm.principal"],
        activityid_activitypointer: ["mscrm.activitypointer"],
        createdby: ["systemuser"],
        createdonbehalfby: ["systemuser"],
        modifiedby: ["systemuser"],
        modifiedonbehalfby: ["systemuser"],
        owningbusinessunit: ["businessunit"],
        owningteam: ["team"],
        owninguser: ["systemuser"],
        regardingobjectid: ["account", "bookableresourcebooking", "bookableresourcebookingheader", "bulkoperation", "campaign", "campaignactivity", "contact", "contract", "entitlement", "entitlementtemplate", "incident", "invoice", "knowledgearticle", "knowledgebaserecord", "lead", "msdyn_customerasset", "msdyn_playbookinstance", "msdyn_postalbum", "msdyn_salessuggestion", "opportunity", "quote", "salesorder", "site"],
        serviceid: ["service"],
        slainvokedid: ["sla"],
        transactioncurrencyid: ["transactioncurrency"],
    },
};
// Attribute constants
var LetterAttributes;
(function (LetterAttributes) {
    LetterAttributes["ActivityId"] = "activityid";
    LetterAttributes["ActivityTypeCode"] = "activitytypecode";
    LetterAttributes["ActualDurationMinutes"] = "actualdurationminutes";
    LetterAttributes["ActualEnd"] = "actualend";
    LetterAttributes["ActualStart"] = "actualstart";
    LetterAttributes["Address"] = "address";
    LetterAttributes["bcc"] = "bcc";
    LetterAttributes["Category"] = "category";
    LetterAttributes["cc"] = "cc";
    LetterAttributes["CreatedBy"] = "createdby";
    LetterAttributes["CreatedByName"] = "createdbyname";
    LetterAttributes["CreatedByYomiName"] = "createdbyyominame";
    LetterAttributes["CreatedOn"] = "createdon";
    LetterAttributes["CreatedOnBehalfBy"] = "createdonbehalfby";
    LetterAttributes["CreatedOnBehalfByName"] = "createdonbehalfbyname";
    LetterAttributes["CreatedOnBehalfByYomiName"] = "createdonbehalfbyyominame";
    LetterAttributes["Description"] = "description";
    LetterAttributes["DirectionCode"] = "directioncode";
    LetterAttributes["ExchangeRate"] = "exchangerate";
    LetterAttributes["from"] = "from";
    LetterAttributes["ImportSequenceNumber"] = "importsequencenumber";
    LetterAttributes["IsBilled"] = "isbilled";
    LetterAttributes["IsRegularActivity"] = "isregularactivity";
    LetterAttributes["IsWorkflowCreated"] = "isworkflowcreated";
    LetterAttributes["LastOnHoldTime"] = "lastonholdtime";
    LetterAttributes["ModifiedBy"] = "modifiedby";
    LetterAttributes["ModifiedByName"] = "modifiedbyname";
    LetterAttributes["ModifiedByYomiName"] = "modifiedbyyominame";
    LetterAttributes["ModifiedOn"] = "modifiedon";
    LetterAttributes["ModifiedOnBehalfBy"] = "modifiedonbehalfby";
    LetterAttributes["ModifiedOnBehalfByName"] = "modifiedonbehalfbyname";
    LetterAttributes["ModifiedOnBehalfByYomiName"] = "modifiedonbehalfbyyominame";
    LetterAttributes["OnHoldTime"] = "onholdtime";
    LetterAttributes["OverriddenCreatedOn"] = "overriddencreatedon";
    LetterAttributes["OwnerId"] = "ownerid";
    LetterAttributes["OwnerIdName"] = "owneridname";
    LetterAttributes["OwnerIdType"] = "owneridtype";
    LetterAttributes["OwnerIdYomiName"] = "owneridyominame";
    LetterAttributes["OwningBusinessUnit"] = "owningbusinessunit";
    LetterAttributes["OwningBusinessUnitName"] = "owningbusinessunitname";
    LetterAttributes["OwningTeam"] = "owningteam";
    LetterAttributes["OwningUser"] = "owninguser";
    LetterAttributes["PriorityCode"] = "prioritycode";
    LetterAttributes["ProcessId"] = "processid";
    LetterAttributes["RegardingObjectId"] = "regardingobjectid";
    LetterAttributes["RegardingObjectIdName"] = "regardingobjectidname";
    LetterAttributes["RegardingObjectIdYomiName"] = "regardingobjectidyominame";
    LetterAttributes["RegardingObjectTypeCode"] = "regardingobjecttypecode";
    LetterAttributes["ScheduledDurationMinutes"] = "scheduleddurationminutes";
    LetterAttributes["ScheduledEnd"] = "scheduledend";
    LetterAttributes["ScheduledStart"] = "scheduledstart";
    LetterAttributes["ServiceId"] = "serviceid";
    LetterAttributes["ServiceIdName"] = "serviceidname";
    LetterAttributes["SLAId"] = "slaid";
    LetterAttributes["SLAInvokedId"] = "slainvokedid";
    LetterAttributes["SLAInvokedIdName"] = "slainvokedidname";
    LetterAttributes["SLAName"] = "slaname";
    LetterAttributes["SortDate"] = "sortdate";
    LetterAttributes["StageId"] = "stageid";
    LetterAttributes["StateCode"] = "statecode";
    LetterAttributes["StatusCode"] = "statuscode";
    LetterAttributes["Subcategory"] = "subcategory";
    LetterAttributes["Subject"] = "subject";
    LetterAttributes["SubscriptionId"] = "subscriptionid";
    LetterAttributes["TimeZoneRuleVersionNumber"] = "timezoneruleversionnumber";
    LetterAttributes["to"] = "to";
    LetterAttributes["TransactionCurrencyId"] = "transactioncurrencyid";
    LetterAttributes["TransactionCurrencyIdName"] = "transactioncurrencyidname";
    LetterAttributes["TraversedPath"] = "traversedpath";
    LetterAttributes["UTCConversionTimeZoneCode"] = "utcconversiontimezonecode";
    LetterAttributes["VersionNumber"] = "versionnumber";
})(LetterAttributes = exports.LetterAttributes || (exports.LetterAttributes = {}));
