// dataverse-ify/lib/dataverse-gen/entities/Opportunity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpportunityAttributes = exports.opportunityMetadata = void 0;
// Entity Opportunity
exports.opportunityMetadata = {
    typeName: "mscrm.opportunity",
    logicalName: "opportunity",
    collectionName: "opportunities",
    primaryIdAttribute: "opportunityid",
    attributeTypes: {
        // Numeric Types
        actualvalue: "Money",
        actualvalue_base: "Money",
        budgetamount: "Money",
        budgetamount_base: "Money",
        closeprobability: "Integer",
        discountamount: "Money",
        discountamount_base: "Money",
        discountpercentage: "Decimal",
        estimatedvalue: "Money",
        estimatedvalue_base: "Money",
        exchangerate: "Decimal",
        freightamount: "Money",
        freightamount_base: "Money",
        importsequencenumber: "Integer",
        msdyn_opportunityscore: "Integer",
        onholdtime: "Integer",
        teamsfollowed: "Integer",
        timezoneruleversionnumber: "Integer",
        totalamount: "Money",
        totalamount_base: "Money",
        totalamountlessfreight: "Money",
        totalamountlessfreight_base: "Money",
        totaldiscountamount: "Money",
        totaldiscountamount_base: "Money",
        totallineitemamount: "Money",
        totallineitemamount_base: "Money",
        totallineitemdiscountamount: "Money",
        totallineitemdiscountamount_base: "Money",
        totaltax: "Money",
        totaltax_base: "Money",
        utcconversiontimezonecode: "Integer",
        versionnumber: "BigInt",
        // Optionsets
        budgetstatus: "Optionset",
        initialcommunication: "Optionset",
        msdyn_forecastcategory: "Optionset",
        msdyn_opportunitygrade: "Optionset",
        msdyn_opportunityscoretrend: "Optionset",
        need: "Optionset",
        opportunityratingcode: "Optionset",
        pricingerrorcode: "Optionset",
        prioritycode: "Optionset",
        purchaseprocess: "Optionset",
        purchasetimeframe: "Optionset",
        salesstage: "Optionset",
        salesstagecode: "Optionset",
        skippricecalculation: "Optionset",
        statecode: "Optionset",
        statuscode: "Optionset",
        timeline: "Optionset",
        // Date Formats
        actualclosedate: "DateOnly:DateOnly",
        createdon: "DateAndTime:UserLocal",
        estimatedclosedate: "DateOnly:DateOnly",
        finaldecisiondate: "DateOnly:DateOnly",
        lastonholdtime: "DateAndTime:UserLocal",
        modifiedon: "DateAndTime:UserLocal",
        overriddencreatedon: "DateOnly:UserLocal",
        schedulefollowup_prospect: "DateOnly:UserLocal",
        schedulefollowup_qualify: "DateOnly:UserLocal",
        scheduleproposalmeeting: "DateOnly:UserLocal",
    },
    navigation: {
        transactioncurrencyid: ["mscrm.transactioncurrency"],
        stageid_processstage: ["mscrm.processstage"],
        sla_opportunity_sla: ["mscrm.sla"],
        pricelevelid: ["mscrm.pricelevel"],
        parentcontactid: ["mscrm.contact"],
        parentaccountid: ["mscrm.account"],
        owninguser: ["mscrm.systemuser"],
        owningteam: ["mscrm.team"],
        owningbusinessunit: ["mscrm.businessunit"],
        ownerid: ["mscrm.principal"],
        originatingleadid: ["mscrm.lead"],
        msdyn_segmentid: ["mscrm.msdyn_segment"],
        msdyn_opportunitykpiid: ["mscrm.msdyn_opportunitykpiitem"],
        msdyn_PredictiveScoreId: ["mscrm.msdyn_predictivescore"],
        modifiedonbehalfby: ["mscrm.systemuser"],
        modifiedby: ["mscrm.systemuser"],
        createdonbehalfby: ["mscrm.systemuser"],
        createdby: ["mscrm.systemuser"],
        campaignid: ["mscrm.campaign"],
        customerid: ["account", "contact"],
        slainvokedid: ["sla"],
    },
};
// Attribute constants
var OpportunityAttributes;
(function (OpportunityAttributes) {
    OpportunityAttributes["AccountId"] = "accountid";
    OpportunityAttributes["AccountIdName"] = "accountidname";
    OpportunityAttributes["AccountIdYomiName"] = "accountidyominame";
    OpportunityAttributes["ActualCloseDate"] = "actualclosedate";
    OpportunityAttributes["ActualValue"] = "actualvalue";
    OpportunityAttributes["ActualValue_Base"] = "actualvalue_base";
    OpportunityAttributes["BudgetAmount"] = "budgetamount";
    OpportunityAttributes["BudgetAmount_Base"] = "budgetamount_base";
    OpportunityAttributes["BudgetStatus"] = "budgetstatus";
    OpportunityAttributes["CampaignId"] = "campaignid";
    OpportunityAttributes["CampaignIdName"] = "campaignidname";
    OpportunityAttributes["CaptureProposalFeedback"] = "captureproposalfeedback";
    OpportunityAttributes["CloseProbability"] = "closeprobability";
    OpportunityAttributes["CompleteFinalProposal"] = "completefinalproposal";
    OpportunityAttributes["CompleteInternalReview"] = "completeinternalreview";
    OpportunityAttributes["ConfirmInterest"] = "confirminterest";
    OpportunityAttributes["ContactId"] = "contactid";
    OpportunityAttributes["ContactIdName"] = "contactidname";
    OpportunityAttributes["ContactIdYomiName"] = "contactidyominame";
    OpportunityAttributes["CreatedBy"] = "createdby";
    OpportunityAttributes["CreatedByName"] = "createdbyname";
    OpportunityAttributes["CreatedByYomiName"] = "createdbyyominame";
    OpportunityAttributes["CreatedOn"] = "createdon";
    OpportunityAttributes["CreatedOnBehalfBy"] = "createdonbehalfby";
    OpportunityAttributes["CreatedOnBehalfByName"] = "createdonbehalfbyname";
    OpportunityAttributes["CreatedOnBehalfByYomiName"] = "createdonbehalfbyyominame";
    OpportunityAttributes["CurrentSituation"] = "currentsituation";
    OpportunityAttributes["CustomerId"] = "customerid";
    OpportunityAttributes["CustomerIdName"] = "customeridname";
    OpportunityAttributes["CustomerIdType"] = "customeridtype";
    OpportunityAttributes["CustomerIdYomiName"] = "customeridyominame";
    OpportunityAttributes["CustomerNeed"] = "customerneed";
    OpportunityAttributes["CustomerPainPoints"] = "customerpainpoints";
    OpportunityAttributes["DecisionMaker"] = "decisionmaker";
    OpportunityAttributes["Description"] = "description";
    OpportunityAttributes["DevelopProposal"] = "developproposal";
    OpportunityAttributes["DiscountAmount"] = "discountamount";
    OpportunityAttributes["DiscountAmount_Base"] = "discountamount_base";
    OpportunityAttributes["DiscountPercentage"] = "discountpercentage";
    OpportunityAttributes["EmailAddress"] = "emailaddress";
    OpportunityAttributes["EstimatedCloseDate"] = "estimatedclosedate";
    OpportunityAttributes["EstimatedValue"] = "estimatedvalue";
    OpportunityAttributes["EstimatedValue_Base"] = "estimatedvalue_base";
    OpportunityAttributes["EvaluateFit"] = "evaluatefit";
    OpportunityAttributes["ExchangeRate"] = "exchangerate";
    OpportunityAttributes["FileDebrief"] = "filedebrief";
    OpportunityAttributes["FinalDecisionDate"] = "finaldecisiondate";
    OpportunityAttributes["FreightAmount"] = "freightamount";
    OpportunityAttributes["FreightAmount_Base"] = "freightamount_base";
    OpportunityAttributes["IdentifyCompetitors"] = "identifycompetitors";
    OpportunityAttributes["IdentifyCustomerContacts"] = "identifycustomercontacts";
    OpportunityAttributes["IdentifyPursuitTeam"] = "identifypursuitteam";
    OpportunityAttributes["ImportSequenceNumber"] = "importsequencenumber";
    OpportunityAttributes["InitialCommunication"] = "initialcommunication";
    OpportunityAttributes["IsPrivate"] = "isprivate";
    OpportunityAttributes["IsRevenueSystemCalculated"] = "isrevenuesystemcalculated";
    OpportunityAttributes["LastOnHoldTime"] = "lastonholdtime";
    OpportunityAttributes["ModifiedBy"] = "modifiedby";
    OpportunityAttributes["ModifiedByName"] = "modifiedbyname";
    OpportunityAttributes["ModifiedByYomiName"] = "modifiedbyyominame";
    OpportunityAttributes["ModifiedOn"] = "modifiedon";
    OpportunityAttributes["ModifiedOnBehalfBy"] = "modifiedonbehalfby";
    OpportunityAttributes["ModifiedOnBehalfByName"] = "modifiedonbehalfbyname";
    OpportunityAttributes["ModifiedOnBehalfByYomiName"] = "modifiedonbehalfbyyominame";
    OpportunityAttributes["msdyn_forecastcategory"] = "msdyn_forecastcategory";
    OpportunityAttributes["msdyn_gdproptout"] = "msdyn_gdproptout";
    OpportunityAttributes["msdyn_OpportunityGrade"] = "msdyn_opportunitygrade";
    OpportunityAttributes["msdyn_opportunitykpiid"] = "msdyn_opportunitykpiid";
    OpportunityAttributes["msdyn_opportunitykpiidName"] = "msdyn_opportunitykpiidname";
    OpportunityAttributes["msdyn_OpportunityScore"] = "msdyn_opportunityscore";
    OpportunityAttributes["msdyn_OpportunityScoreTrend"] = "msdyn_opportunityscoretrend";
    OpportunityAttributes["msdyn_PredictiveScoreId"] = "msdyn_predictivescoreid";
    OpportunityAttributes["msdyn_PredictiveScoreIdName"] = "msdyn_predictivescoreidname";
    OpportunityAttributes["msdyn_ScoreHistory"] = "msdyn_scorehistory";
    OpportunityAttributes["msdyn_ScoreReasons"] = "msdyn_scorereasons";
    OpportunityAttributes["msdyn_segmentid"] = "msdyn_segmentid";
    OpportunityAttributes["msdyn_segmentidName"] = "msdyn_segmentidname";
    OpportunityAttributes["msdyn_similaropportunities"] = "msdyn_similaropportunities";
    OpportunityAttributes["Name"] = "name";
    OpportunityAttributes["Need"] = "need";
    OpportunityAttributes["OnHoldTime"] = "onholdtime";
    OpportunityAttributes["OpportunityId"] = "opportunityid";
    OpportunityAttributes["OpportunityRatingCode"] = "opportunityratingcode";
    OpportunityAttributes["OriginatingLeadId"] = "originatingleadid";
    OpportunityAttributes["OriginatingLeadIdName"] = "originatingleadidname";
    OpportunityAttributes["OriginatingLeadIdYomiName"] = "originatingleadidyominame";
    OpportunityAttributes["OverriddenCreatedOn"] = "overriddencreatedon";
    OpportunityAttributes["OwnerId"] = "ownerid";
    OpportunityAttributes["OwnerIdName"] = "owneridname";
    OpportunityAttributes["OwnerIdType"] = "owneridtype";
    OpportunityAttributes["OwnerIdYomiName"] = "owneridyominame";
    OpportunityAttributes["OwningBusinessUnit"] = "owningbusinessunit";
    OpportunityAttributes["OwningBusinessUnitName"] = "owningbusinessunitname";
    OpportunityAttributes["OwningTeam"] = "owningteam";
    OpportunityAttributes["OwningUser"] = "owninguser";
    OpportunityAttributes["ParentAccountId"] = "parentaccountid";
    OpportunityAttributes["ParentAccountIdName"] = "parentaccountidname";
    OpportunityAttributes["ParentAccountIdYomiName"] = "parentaccountidyominame";
    OpportunityAttributes["ParentContactId"] = "parentcontactid";
    OpportunityAttributes["ParentContactIdName"] = "parentcontactidname";
    OpportunityAttributes["ParentContactIdYomiName"] = "parentcontactidyominame";
    OpportunityAttributes["ParticipatesInWorkflow"] = "participatesinworkflow";
    OpportunityAttributes["PresentFinalProposal"] = "presentfinalproposal";
    OpportunityAttributes["PresentProposal"] = "presentproposal";
    OpportunityAttributes["PriceLevelId"] = "pricelevelid";
    OpportunityAttributes["PriceLevelIdName"] = "pricelevelidname";
    OpportunityAttributes["PricingErrorCode"] = "pricingerrorcode";
    OpportunityAttributes["PriorityCode"] = "prioritycode";
    OpportunityAttributes["ProcessId"] = "processid";
    OpportunityAttributes["ProposedSolution"] = "proposedsolution";
    OpportunityAttributes["PurchaseProcess"] = "purchaseprocess";
    OpportunityAttributes["PurchaseTimeframe"] = "purchasetimeframe";
    OpportunityAttributes["PursuitDecision"] = "pursuitdecision";
    OpportunityAttributes["QualificationComments"] = "qualificationcomments";
    OpportunityAttributes["QuoteComments"] = "quotecomments";
    OpportunityAttributes["ResolveFeedback"] = "resolvefeedback";
    OpportunityAttributes["SalesStage"] = "salesstage";
    OpportunityAttributes["SalesStageCode"] = "salesstagecode";
    OpportunityAttributes["ScheduleFollowup_Prospect"] = "schedulefollowup_prospect";
    OpportunityAttributes["ScheduleFollowup_Qualify"] = "schedulefollowup_qualify";
    OpportunityAttributes["ScheduleProposalMeeting"] = "scheduleproposalmeeting";
    OpportunityAttributes["SendThankYouNote"] = "sendthankyounote";
    OpportunityAttributes["SkipPriceCalculation"] = "skippricecalculation";
    OpportunityAttributes["SLAId"] = "slaid";
    OpportunityAttributes["SLAInvokedId"] = "slainvokedid";
    OpportunityAttributes["SLAInvokedIdName"] = "slainvokedidname";
    OpportunityAttributes["SLAName"] = "slaname";
    OpportunityAttributes["StageId"] = "stageid";
    OpportunityAttributes["StateCode"] = "statecode";
    OpportunityAttributes["StatusCode"] = "statuscode";
    OpportunityAttributes["StepId"] = "stepid";
    OpportunityAttributes["StepName"] = "stepname";
    OpportunityAttributes["TeamsFollowed"] = "teamsfollowed";
    OpportunityAttributes["TimeLine"] = "timeline";
    OpportunityAttributes["TimeSpentByMeOnEmailAndMeetings"] = "timespentbymeonemailandmeetings";
    OpportunityAttributes["TimeZoneRuleVersionNumber"] = "timezoneruleversionnumber";
    OpportunityAttributes["TotalAmount"] = "totalamount";
    OpportunityAttributes["TotalAmount_Base"] = "totalamount_base";
    OpportunityAttributes["TotalAmountLessFreight"] = "totalamountlessfreight";
    OpportunityAttributes["TotalAmountLessFreight_Base"] = "totalamountlessfreight_base";
    OpportunityAttributes["TotalDiscountAmount"] = "totaldiscountamount";
    OpportunityAttributes["TotalDiscountAmount_Base"] = "totaldiscountamount_base";
    OpportunityAttributes["TotalLineItemAmount"] = "totallineitemamount";
    OpportunityAttributes["TotalLineItemAmount_Base"] = "totallineitemamount_base";
    OpportunityAttributes["TotalLineItemDiscountAmount"] = "totallineitemdiscountamount";
    OpportunityAttributes["TotalLineItemDiscountAmount_Base"] = "totallineitemdiscountamount_base";
    OpportunityAttributes["TotalTax"] = "totaltax";
    OpportunityAttributes["TotalTax_Base"] = "totaltax_base";
    OpportunityAttributes["TransactionCurrencyId"] = "transactioncurrencyid";
    OpportunityAttributes["TransactionCurrencyIdName"] = "transactioncurrencyidname";
    OpportunityAttributes["TraversedPath"] = "traversedpath";
    OpportunityAttributes["UTCConversionTimeZoneCode"] = "utcconversiontimezonecode";
    OpportunityAttributes["VersionNumber"] = "versionnumber";
})(OpportunityAttributes = exports.OpportunityAttributes || (exports.OpportunityAttributes = {}));
