// dataverse-ify/lib/dataverse-gen/entities/OpportunityClose.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpportunityCloseAttributes = exports.opportunitycloseMetadata = void 0;
// Entity OpportunityClose
exports.opportunitycloseMetadata = {
    typeName: "mscrm.opportunityclose",
    logicalName: "opportunityclose",
    collectionName: "opportunitycloses",
    primaryIdAttribute: "activityid",
    attributeTypes: {
        // Numeric Types
        actualdurationminutes: "Integer",
        actualrevenue: "Money",
        actualrevenue_base: "Money",
        exchangerate: "Decimal",
        importsequencenumber: "Integer",
        onholdtime: "Integer",
        scheduleddurationminutes: "Integer",
        timezoneruleversionnumber: "Integer",
        utcconversiontimezonecode: "Integer",
        versionnumber: "BigInt",
        // Optionsets
        community: "Optionset",
        deliveryprioritycode: "Optionset",
        instancetypecode: "Optionset",
        opportunitystatecode: "Optionset",
        opportunitystatuscode: "Optionset",
        prioritycode: "Optionset",
        statecode: "Optionset",
        statuscode: "Optionset",
        // Date Formats
        actualend: "DateOnly:UserLocal",
        actualstart: "DateOnly:UserLocal",
        createdon: "DateAndTime:UserLocal",
        deliverylastattemptedon: "DateAndTime:UserLocal",
        lastonholdtime: "DateAndTime:UserLocal",
        modifiedon: "DateAndTime:UserLocal",
        overriddencreatedon: "DateOnly:UserLocal",
        postponeactivityprocessinguntil: "DateAndTime:UserLocal",
        scheduledend: "DateOnly:UserLocal",
        scheduledstart: "DateOnly:UserLocal",
        senton: "DateAndTime:UserLocal",
        sortdate: "DateAndTime:UserLocal",
    },
    navigation: {
        sla_activitypointer_sla_opportunityclose: ["mscrm.sla"],
        ownerid_opportunityclose: ["mscrm.principal"],
        opportunityid: ["mscrm.opportunity"],
        competitorid: ["mscrm.competitor"],
        activityid_activitypointer: ["mscrm.activitypointer"],
        createdby: ["systemuser"],
        createdonbehalfby: ["systemuser"],
        modifiedby: ["systemuser"],
        modifiedonbehalfby: ["systemuser"],
        owningbusinessunit: ["businessunit"],
        owningteam: ["team"],
        owninguser: ["systemuser"],
        regardingobjectid: ["account", "bookableresourcebooking", "bookableresourcebookingheader", "bulkoperation", "campaign", "campaignactivity", "entitlement", "entitlementtemplate", "incident", "interactionforemail", "knowledgearticle", "knowledgebaserecord", "lead", "msdyn_customerasset", "msdyn_playbookinstance", "opportunity", "site"],
        sendermailboxid: ["mailbox"],
        serviceid: ["service"],
        slainvokedid: ["sla"],
        transactioncurrencyid: ["transactioncurrency"],
    },
};
// Attribute constants
var OpportunityCloseAttributes;
(function (OpportunityCloseAttributes) {
    OpportunityCloseAttributes["ActivityAdditionalParams"] = "activityadditionalparams";
    OpportunityCloseAttributes["ActivityId"] = "activityid";
    OpportunityCloseAttributes["ActivityTypeCode"] = "activitytypecode";
    OpportunityCloseAttributes["ActualDurationMinutes"] = "actualdurationminutes";
    OpportunityCloseAttributes["ActualEnd"] = "actualend";
    OpportunityCloseAttributes["ActualRevenue"] = "actualrevenue";
    OpportunityCloseAttributes["ActualRevenue_Base"] = "actualrevenue_base";
    OpportunityCloseAttributes["ActualStart"] = "actualstart";
    OpportunityCloseAttributes["BCC"] = "bcc";
    OpportunityCloseAttributes["Category"] = "category";
    OpportunityCloseAttributes["CC"] = "cc";
    OpportunityCloseAttributes["Community"] = "community";
    OpportunityCloseAttributes["CompetitorId"] = "competitorid";
    OpportunityCloseAttributes["CompetitorIdName"] = "competitoridname";
    OpportunityCloseAttributes["CompetitorIdYomiName"] = "competitoridyominame";
    OpportunityCloseAttributes["CreatedBy"] = "createdby";
    OpportunityCloseAttributes["CreatedByExternalParty"] = "createdbyexternalparty";
    OpportunityCloseAttributes["CreatedByExternalPartyName"] = "createdbyexternalpartyname";
    OpportunityCloseAttributes["CreatedByExternalPartyYomiName"] = "createdbyexternalpartyyominame";
    OpportunityCloseAttributes["CreatedByName"] = "createdbyname";
    OpportunityCloseAttributes["CreatedByYomiName"] = "createdbyyominame";
    OpportunityCloseAttributes["CreatedOn"] = "createdon";
    OpportunityCloseAttributes["CreatedOnBehalfBy"] = "createdonbehalfby";
    OpportunityCloseAttributes["CreatedOnBehalfByName"] = "createdonbehalfbyname";
    OpportunityCloseAttributes["CreatedOnBehalfByYomiName"] = "createdonbehalfbyyominame";
    OpportunityCloseAttributes["Customers"] = "customers";
    OpportunityCloseAttributes["DeliveryLastAttemptedOn"] = "deliverylastattemptedon";
    OpportunityCloseAttributes["DeliveryPriorityCode"] = "deliveryprioritycode";
    OpportunityCloseAttributes["Description"] = "description";
    OpportunityCloseAttributes["ExchangeItemId"] = "exchangeitemid";
    OpportunityCloseAttributes["ExchangeRate"] = "exchangerate";
    OpportunityCloseAttributes["ExchangeWebLink"] = "exchangeweblink";
    OpportunityCloseAttributes["From"] = "from";
    OpportunityCloseAttributes["ImportSequenceNumber"] = "importsequencenumber";
    OpportunityCloseAttributes["InstanceTypeCode"] = "instancetypecode";
    OpportunityCloseAttributes["IsBilled"] = "isbilled";
    OpportunityCloseAttributes["IsMapiPrivate"] = "ismapiprivate";
    OpportunityCloseAttributes["IsRegularActivity"] = "isregularactivity";
    OpportunityCloseAttributes["IsWorkflowCreated"] = "isworkflowcreated";
    OpportunityCloseAttributes["LastOnHoldTime"] = "lastonholdtime";
    OpportunityCloseAttributes["LeftVoiceMail"] = "leftvoicemail";
    OpportunityCloseAttributes["ModifiedBy"] = "modifiedby";
    OpportunityCloseAttributes["ModifiedByExternalParty"] = "modifiedbyexternalparty";
    OpportunityCloseAttributes["ModifiedByExternalPartyName"] = "modifiedbyexternalpartyname";
    OpportunityCloseAttributes["ModifiedByExternalPartyYomiName"] = "modifiedbyexternalpartyyominame";
    OpportunityCloseAttributes["ModifiedByName"] = "modifiedbyname";
    OpportunityCloseAttributes["ModifiedByYomiName"] = "modifiedbyyominame";
    OpportunityCloseAttributes["ModifiedOn"] = "modifiedon";
    OpportunityCloseAttributes["ModifiedOnBehalfBy"] = "modifiedonbehalfby";
    OpportunityCloseAttributes["ModifiedOnBehalfByName"] = "modifiedonbehalfbyname";
    OpportunityCloseAttributes["ModifiedOnBehalfByYomiName"] = "modifiedonbehalfbyyominame";
    OpportunityCloseAttributes["OnHoldTime"] = "onholdtime";
    OpportunityCloseAttributes["OpportunityId"] = "opportunityid";
    OpportunityCloseAttributes["OpportunityIdName"] = "opportunityidname";
    OpportunityCloseAttributes["OpportunityIdType"] = "opportunityidtype";
    OpportunityCloseAttributes["OpportunityStateCode"] = "opportunitystatecode";
    OpportunityCloseAttributes["OpportunityStatusCode"] = "opportunitystatuscode";
    OpportunityCloseAttributes["OptionalAttendees"] = "optionalattendees";
    OpportunityCloseAttributes["Organizer"] = "organizer";
    OpportunityCloseAttributes["OverriddenCreatedOn"] = "overriddencreatedon";
    OpportunityCloseAttributes["OwnerId"] = "ownerid";
    OpportunityCloseAttributes["OwnerIdName"] = "owneridname";
    OpportunityCloseAttributes["OwnerIdType"] = "owneridtype";
    OpportunityCloseAttributes["OwnerIdYomiName"] = "owneridyominame";
    OpportunityCloseAttributes["OwningBusinessUnit"] = "owningbusinessunit";
    OpportunityCloseAttributes["OwningBusinessUnitName"] = "owningbusinessunitname";
    OpportunityCloseAttributes["OwningTeam"] = "owningteam";
    OpportunityCloseAttributes["OwningUser"] = "owninguser";
    OpportunityCloseAttributes["Partners"] = "partners";
    OpportunityCloseAttributes["PostponeActivityProcessingUntil"] = "postponeactivityprocessinguntil";
    OpportunityCloseAttributes["PriorityCode"] = "prioritycode";
    OpportunityCloseAttributes["ProcessId"] = "processid";
    OpportunityCloseAttributes["RegardingObjectId"] = "regardingobjectid";
    OpportunityCloseAttributes["RegardingObjectIdName"] = "regardingobjectidname";
    OpportunityCloseAttributes["RegardingObjectIdYomiName"] = "regardingobjectidyominame";
    OpportunityCloseAttributes["RegardingObjectTypeCode"] = "regardingobjecttypecode";
    OpportunityCloseAttributes["RequiredAttendees"] = "requiredattendees";
    OpportunityCloseAttributes["Resources"] = "resources";
    OpportunityCloseAttributes["ScheduledDurationMinutes"] = "scheduleddurationminutes";
    OpportunityCloseAttributes["ScheduledEnd"] = "scheduledend";
    OpportunityCloseAttributes["ScheduledStart"] = "scheduledstart";
    OpportunityCloseAttributes["SenderMailboxId"] = "sendermailboxid";
    OpportunityCloseAttributes["SenderMailboxIdName"] = "sendermailboxidname";
    OpportunityCloseAttributes["SentOn"] = "senton";
    OpportunityCloseAttributes["SeriesId"] = "seriesid";
    OpportunityCloseAttributes["ServiceId"] = "serviceid";
    OpportunityCloseAttributes["ServiceIdName"] = "serviceidname";
    OpportunityCloseAttributes["SLAId"] = "slaid";
    OpportunityCloseAttributes["SLAInvokedId"] = "slainvokedid";
    OpportunityCloseAttributes["SLAInvokedIdName"] = "slainvokedidname";
    OpportunityCloseAttributes["SLAName"] = "slaname";
    OpportunityCloseAttributes["SortDate"] = "sortdate";
    OpportunityCloseAttributes["StageId"] = "stageid";
    OpportunityCloseAttributes["StateCode"] = "statecode";
    OpportunityCloseAttributes["StatusCode"] = "statuscode";
    OpportunityCloseAttributes["Subcategory"] = "subcategory";
    OpportunityCloseAttributes["Subject"] = "subject";
    OpportunityCloseAttributes["TimeZoneRuleVersionNumber"] = "timezoneruleversionnumber";
    OpportunityCloseAttributes["To"] = "to";
    OpportunityCloseAttributes["TransactionCurrencyId"] = "transactioncurrencyid";
    OpportunityCloseAttributes["TransactionCurrencyIdName"] = "transactioncurrencyidname";
    OpportunityCloseAttributes["TraversedPath"] = "traversedpath";
    OpportunityCloseAttributes["UTCConversionTimeZoneCode"] = "utcconversiontimezonecode";
    OpportunityCloseAttributes["VersionNumber"] = "versionnumber";
})(OpportunityCloseAttributes = exports.OpportunityCloseAttributes || (exports.OpportunityCloseAttributes = {}));
