// dataverse-ify/lib/dataverse-gen/entities/Queue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueueAttributes = exports.queueMetadata = void 0;
// Entity Queue
exports.queueMetadata = {
    typeName: "mscrm.queue",
    logicalName: "queue",
    collectionName: "queues",
    primaryIdAttribute: "queueid",
    attributeTypes: {
        // Numeric Types
        entityimage_timestamp: "BigInt",
        exchangerate: "Decimal",
        importsequencenumber: "Integer",
        numberofitems: "Integer",
        numberofmembers: "Integer",
        versionnumber: "BigInt",
        // Optionsets
        emailrouteraccessapproval: "Optionset",
        incomingemaildeliverymethod: "Optionset",
        incomingemailfilteringmethod: "Optionset",
        outgoingemaildeliverymethod: "Optionset",
        queuetypecode: "Optionset",
        queueviewtype: "Optionset",
        statecode: "Optionset",
        statuscode: "Optionset",
        // Date Formats
        createdon: "DateAndTime:UserLocal",
        modifiedon: "DateAndTime:UserLocal",
        overriddencreatedon: "DateOnly:UserLocal",
    },
    navigation: {
        transactioncurrencyid: ["mscrm.transactioncurrency"],
        primaryuserid: ["mscrm.systemuser"],
        owningbusinessunit: ["mscrm.businessunit"],
        ownerid: ["mscrm.principal"],
        organizationid: ["mscrm.organization"],
        modifiedonbehalfby: ["mscrm.systemuser"],
        modifiedby: ["mscrm.systemuser"],
        defaultmailbox: ["mscrm.mailbox"],
        createdonbehalfby: ["mscrm.systemuser"],
        createdby: ["mscrm.systemuser"],
        businessunitid: ["mscrm.businessunit"],
    },
};
// Attribute constants
var QueueAttributes;
(function (QueueAttributes) {
    QueueAttributes["AllowEmailCredentials"] = "allowemailcredentials";
    QueueAttributes["BusinessUnitId"] = "businessunitid";
    QueueAttributes["BusinessUnitIdName"] = "businessunitidname";
    QueueAttributes["CreatedBy"] = "createdby";
    QueueAttributes["CreatedByName"] = "createdbyname";
    QueueAttributes["CreatedByYomiName"] = "createdbyyominame";
    QueueAttributes["CreatedOn"] = "createdon";
    QueueAttributes["CreatedOnBehalfBy"] = "createdonbehalfby";
    QueueAttributes["CreatedOnBehalfByName"] = "createdonbehalfbyname";
    QueueAttributes["CreatedOnBehalfByYomiName"] = "createdonbehalfbyyominame";
    QueueAttributes["DefaultMailbox"] = "defaultmailbox";
    QueueAttributes["DefaultMailboxName"] = "defaultmailboxname";
    QueueAttributes["Description"] = "description";
    QueueAttributes["EMailAddress"] = "emailaddress";
    QueueAttributes["EmailPassword"] = "emailpassword";
    QueueAttributes["EmailRouterAccessApproval"] = "emailrouteraccessapproval";
    QueueAttributes["EmailUsername"] = "emailusername";
    QueueAttributes["EntityImage"] = "entityimage";
    QueueAttributes["EntityImage_Timestamp"] = "entityimage_timestamp";
    QueueAttributes["EntityImage_URL"] = "entityimage_url";
    QueueAttributes["EntityImageId"] = "entityimageid";
    QueueAttributes["ExchangeRate"] = "exchangerate";
    QueueAttributes["IgnoreUnsolicitedEmail"] = "ignoreunsolicitedemail";
    QueueAttributes["ImportSequenceNumber"] = "importsequencenumber";
    QueueAttributes["IncomingEmailDeliveryMethod"] = "incomingemaildeliverymethod";
    QueueAttributes["IncomingEmailFilteringMethod"] = "incomingemailfilteringmethod";
    QueueAttributes["IsEmailAddressApprovedByO365Admin"] = "isemailaddressapprovedbyo365admin";
    QueueAttributes["IsFaxQueue"] = "isfaxqueue";
    QueueAttributes["ModifiedBy"] = "modifiedby";
    QueueAttributes["ModifiedByName"] = "modifiedbyname";
    QueueAttributes["ModifiedByYomiName"] = "modifiedbyyominame";
    QueueAttributes["ModifiedOn"] = "modifiedon";
    QueueAttributes["ModifiedOnBehalfBy"] = "modifiedonbehalfby";
    QueueAttributes["ModifiedOnBehalfByName"] = "modifiedonbehalfbyname";
    QueueAttributes["ModifiedOnBehalfByYomiName"] = "modifiedonbehalfbyyominame";
    QueueAttributes["Name"] = "name";
    QueueAttributes["NumberOfItems"] = "numberofitems";
    QueueAttributes["NumberOfMembers"] = "numberofmembers";
    QueueAttributes["OrganizationId"] = "organizationid";
    QueueAttributes["OrganizationIdName"] = "organizationidname";
    QueueAttributes["OutgoingEmailDeliveryMethod"] = "outgoingemaildeliverymethod";
    QueueAttributes["OverriddenCreatedOn"] = "overriddencreatedon";
    QueueAttributes["OwnerId"] = "ownerid";
    QueueAttributes["OwnerIdName"] = "owneridname";
    QueueAttributes["OwnerIdType"] = "owneridtype";
    QueueAttributes["OwnerIdYomiName"] = "owneridyominame";
    QueueAttributes["OwningBusinessUnit"] = "owningbusinessunit";
    QueueAttributes["OwningBusinessUnitName"] = "owningbusinessunitname";
    QueueAttributes["OwningTeam"] = "owningteam";
    QueueAttributes["OwningUser"] = "owninguser";
    QueueAttributes["PrimaryUserId"] = "primaryuserid";
    QueueAttributes["PrimaryUserIdName"] = "primaryuseridname";
    QueueAttributes["PrimaryUserIdYomiName"] = "primaryuseridyominame";
    QueueAttributes["QueueId"] = "queueid";
    QueueAttributes["QueueTypeCode"] = "queuetypecode";
    QueueAttributes["QueueViewType"] = "queueviewtype";
    QueueAttributes["StateCode"] = "statecode";
    QueueAttributes["StatusCode"] = "statuscode";
    QueueAttributes["TransactionCurrencyId"] = "transactioncurrencyid";
    QueueAttributes["TransactionCurrencyIdName"] = "transactioncurrencyidname";
    QueueAttributes["VersionNumber"] = "versionnumber";
})(QueueAttributes = exports.QueueAttributes || (exports.QueueAttributes = {}));
