// dataverse-ify/lib/dataverse-gen/entities/QueueItem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueueItemAttributes = exports.queueitemMetadata = void 0;
// Entity QueueItem
exports.queueitemMetadata = {
    typeName: "mscrm.queueitem",
    logicalName: "queueitem",
    collectionName: "queueitems",
    primaryIdAttribute: "queueitemid",
    attributeTypes: {
        // Numeric Types
        exchangerate: "Decimal",
        importsequencenumber: "Integer",
        priority: "Integer",
        state: "Integer",
        status: "Integer",
        timezoneruleversionnumber: "Integer",
        utcconversiontimezonecode: "Integer",
        versionnumber: "BigInt",
        // Optionsets
        objecttypecode: "Optionset",
        statecode: "Optionset",
        statuscode: "Optionset",
        // Date Formats
        createdon: "DateAndTime:UserLocal",
        enteredon: "DateAndTime:UserLocal",
        modifiedon: "DateAndTime:UserLocal",
        overriddencreatedon: "DateOnly:UserLocal",
        workeridmodifiedon: "DateOnly:UserLocal",
    },
    navigation: {
        transactioncurrencyid: ["mscrm.transactioncurrency"],
        queueid: ["mscrm.queue"],
        organizationid: ["mscrm.organization"],
        modifiedonbehalfby: ["mscrm.systemuser"],
        modifiedby: ["mscrm.systemuser"],
        createdonbehalfby: ["mscrm.systemuser"],
        createdby: ["mscrm.systemuser"],
        objectid: ["activitypointer", "appointment", "bulkoperation", "campaignactivity", "campaignresponse", "chat", "email", "fax", "incident", "knowledgearticle", "letter", "msdyn_iotalert", "msdyn_knowledgearticletemplate", "msfp_alert", "msfp_surveyinvite", "msfp_surveyresponse", "phonecall", "recurringappointmentmaster", "serviceappointment", "socialactivity", "task"],
        workerid: ["systemuser", "team"],
    },
};
// Attribute constants
var QueueItemAttributes;
(function (QueueItemAttributes) {
    QueueItemAttributes["CreatedBy"] = "createdby";
    QueueItemAttributes["CreatedByName"] = "createdbyname";
    QueueItemAttributes["CreatedByYomiName"] = "createdbyyominame";
    QueueItemAttributes["CreatedOn"] = "createdon";
    QueueItemAttributes["CreatedOnBehalfBy"] = "createdonbehalfby";
    QueueItemAttributes["CreatedOnBehalfByName"] = "createdonbehalfbyname";
    QueueItemAttributes["CreatedOnBehalfByYomiName"] = "createdonbehalfbyyominame";
    QueueItemAttributes["EnteredOn"] = "enteredon";
    QueueItemAttributes["ExchangeRate"] = "exchangerate";
    QueueItemAttributes["ImportSequenceNumber"] = "importsequencenumber";
    QueueItemAttributes["ModifiedBy"] = "modifiedby";
    QueueItemAttributes["ModifiedByName"] = "modifiedbyname";
    QueueItemAttributes["ModifiedByYomiName"] = "modifiedbyyominame";
    QueueItemAttributes["ModifiedOn"] = "modifiedon";
    QueueItemAttributes["ModifiedOnBehalfBy"] = "modifiedonbehalfby";
    QueueItemAttributes["ModifiedOnBehalfByName"] = "modifiedonbehalfbyname";
    QueueItemAttributes["ModifiedOnBehalfByYomiName"] = "modifiedonbehalfbyyominame";
    QueueItemAttributes["ObjectId"] = "objectid";
    QueueItemAttributes["ObjectIdName"] = "objectidname";
    QueueItemAttributes["ObjectIdTypeCode"] = "objectidtypecode";
    QueueItemAttributes["ObjectTypeCode"] = "objecttypecode";
    QueueItemAttributes["OrganizationId"] = "organizationid";
    QueueItemAttributes["OrganizationIdName"] = "organizationidname";
    QueueItemAttributes["OverriddenCreatedOn"] = "overriddencreatedon";
    QueueItemAttributes["OwnerId"] = "ownerid";
    QueueItemAttributes["OwnerIdType"] = "owneridtype";
    QueueItemAttributes["OwningBusinessUnit"] = "owningbusinessunit";
    QueueItemAttributes["OwningUser"] = "owninguser";
    QueueItemAttributes["Priority"] = "priority";
    QueueItemAttributes["QueueId"] = "queueid";
    QueueItemAttributes["QueueIdName"] = "queueidname";
    QueueItemAttributes["QueueItemId"] = "queueitemid";
    QueueItemAttributes["Sender"] = "sender";
    QueueItemAttributes["State"] = "state";
    QueueItemAttributes["StateCode"] = "statecode";
    QueueItemAttributes["Status"] = "status";
    QueueItemAttributes["StatusCode"] = "statuscode";
    QueueItemAttributes["TimeZoneRuleVersionNumber"] = "timezoneruleversionnumber";
    QueueItemAttributes["Title"] = "title";
    QueueItemAttributes["ToRecipients"] = "torecipients";
    QueueItemAttributes["TransactionCurrencyId"] = "transactioncurrencyid";
    QueueItemAttributes["TransactionCurrencyIdName"] = "transactioncurrencyidname";
    QueueItemAttributes["UTCConversionTimeZoneCode"] = "utcconversiontimezonecode";
    QueueItemAttributes["VersionNumber"] = "versionnumber";
    QueueItemAttributes["WorkerId"] = "workerid";
    QueueItemAttributes["WorkerIdModifiedOn"] = "workeridmodifiedon";
    QueueItemAttributes["WorkerIdName"] = "workeridname";
    QueueItemAttributes["WorkerIdType"] = "workeridtype";
    QueueItemAttributes["WorkerIdYomiName"] = "workeridyominame";
})(QueueItemAttributes = exports.QueueItemAttributes || (exports.QueueItemAttributes = {}));
