// dataverse-ify/lib/dataverse-gen/actions/AddToQueue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AddToQueueMetadata = void 0;
const StructuralProperty_1 = require("../../types/StructuralProperty");
const OperationType_1 = require("../../types/OperationType");
// Action AddToQueue
exports.AddToQueueMetadata = {
    boundParameter: "entity",
    parameterTypes: {
        "entity": {
            typeName: "mscrm.queue",
            structuralProperty: StructuralProperty_1.StructuralProperty.EntityType
        },
        "Target": {
            typeName: "mscrm.crmbaseentity",
            structuralProperty: StructuralProperty_1.StructuralProperty.EntityType
        },
        "SourceQueue": {
            typeName: "mscrm.queue",
            structuralProperty: StructuralProperty_1.StructuralProperty.EntityType
        },
        "QueueItemProperties": {
            typeName: "mscrm.queueitem",
            structuralProperty: StructuralProperty_1.StructuralProperty.EntityType
        },
    },
    operationType: OperationType_1.OperationType.Action,
    operationName: "AddToQueue"
};
