// dataverse-ify/lib/webapi/__tests__/CRUD/Create.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable @typescript-eslint/no-explicit-any */
const SetupGlobalContext_1 = require("../../SetupGlobalContext");
const whoAmI_1 = require("../../whoAmI");
describe("XrmWebApiNode CRUD", () => {
    beforeAll(() => __awaiter(void 0, void 0, void 0, function* () {
        // Is this running inside NodeJS?
        if (typeof Xrm === "undefined") {
            try {
                // Set up the Node Xrm global context
                yield SetupGlobalContext_1.SetupGlobalContext();
            }
            catch (ex) {
                fail(ex);
            }
        }
    }), 10000);
    test("Create, Update then Delete", () => __awaiter(void 0, void 0, void 0, function* () {
        const userId = yield whoAmI_1.whoAmI();
        const account1 = {
            name: "Sample Account",
            accountcategorycode: 1,
            creditlimit: 1000,
            creditonhold: true,
            numberofemployees: 10,
            lastonholdtime: new Date(),
            "preferredsystemuserid@odata.bind": `systemusers(${userId})`,
        };
        try {
            // Create Account
            const newAccount = (yield Xrm.WebApi.createRecord("account", account1));
            account1.accountid = newAccount.id;
            if (!account1.accountid) {
                throw new Error("Account not created");
            }
            // Check the account has been created
            const account2 = (yield Xrm.WebApi.retrieveRecord("account", account1.accountid, "?$select=name"));
            expect(account2.name).toBe("Sample Account");
        }
        finally {
            // Delete account
            if (account1.accountid) {
                yield Xrm.WebApi.deleteRecord("account", account1.accountid);
            }
        }
    }), 10000);
});
